# **League of Legends. Этап 1 - Extract. Извлечение данных с использованием официального API Riot Games**

15.06.2026

## Цель:

1. Реализовать	сборщик	данных,	который собирает	игроков	из	лиг	Challenger,	Grandmaster,	Master	для	
регионов	euw1	и	na1.	
2.  Для	каждого	игрока	загружает	до	100	идентификаторов	его	рейтинговых	матчей	за	текущий	месяц.	
3. По	каждому	матчу	загружает	детальную	информацию	(участники,	убийства,	смерти,	вспоможения,	победа/поражение).	

Учитывать	rate	limits	API:	20	запросов	в	секунду,	100	запросов	за	120	секунд.	

Реализовать	планировщик	с	очередью	времени.

## Импорт/установка библиотек

In [2]:
# Устанавливаем  библиотеки: для отправки HTTP-запросов (нужна для API и парсинга сайтов)
#!chcp 65001
#pip install -q requests 
#pip install python-dotenv

In [3]:
import os
from dotenv import load_dotenv
from pathlib import Path
import logging # для логирования

In [4]:
import gc # Garbage Collector Сборщик мусора для очистки памяти
import json
import requests   # для HTTP-запросов к API
import pandas as pd  # для работы с таблицами
import time          # чтобы делать паузы между запросами и не получить бан от API
from datetime import datetime, timedelta, timezone

## Предварительные настройки

In [6]:
# базовый путь
base_dir = Path(os.getcwd()) 

In [7]:
# Загрузка личного токена из файла настройки окружения lol.env
env_file = base_dir / "lol.env"
if env_file.is_file():
    # Загружаем токен, передавая полный путь
    load_dotenv(env_file)
    display("Файл успешно найден и загружен!")
    RIOT_API_KEY = os.getenv("RIOT_API_KEY")
else:
    display(f"❌ Ошибка: Файла нет в папке {base_dir}.")

# Выводим результат
display(f"Токен Riot API состоит из : {len(RIOT_API_KEY)} символов") 

'Файл успешно найден и загружен!'

'Токен Riot API состоит из : 42 символов'

In [8]:
# закголовок для запроса
HEADERS = {"X-Riot-Token": RIOT_API_KEY}

In [7]:
# Список регионов для сбора: euw1 - сервер Europe West (Западная Европа), na1 - North America (Северная Америка)
array_regions = ["euw1", "na1"]  

In [8]:
# Список лиг 
array_leagues = ["challenger", "grandmaster", 'master']  

In [9]:
# Очередь - из какой турнирной таблицы (очереди) нужно забрать список игроков. 
# RANKED_SOLO_5x5 - самый популярный и престижный соревновательный режим, где игроки заходят в матч в одиночку или вдвоем с другом.
QUEUE = "RANKED_SOLO_5x5"

In [10]:
def get_route(region):
    """ Функция для определения глобального маршрута (ROUTE) по региону, т.к. в некоторых эндпоинтах требуется полное название сервера 
    """
    if region =="na1":
        return "americas" 
    else: return "europe"

## Сбор игроков из лиг Challenger, Grandmaster, Master для регионов euw1 и na1

Рассматривемые лиги:
- **Challenger — Лига Претендентов**
- **Grandmaster — Лига Грандмастеров**
- **Master — Лига Мастеров**
  
**GET /lol/league/v4/challengerleagues/by-queue/{queue}** возвращает игроков только из лиги Challenger. Ввозвращает объект типа LeagueListDTO - метаданные лиги конкретного региона и массив всех топ-игроков - список лучших игроков сервера (до 300 человек в зависимости от региона).

**GET /lol/league/v4/grandmasterleagues/by-queue/{queue}** — возвращает игроков для лиги Grandmaster.

**GET /lol/league/v4/masterleagues/by-queue/{queue}** — возвращает игроков для лиги Master.

Структура ответов для всех трёх лиг абсолютно одинаковая. Сервер Riot Games возвращает идентичный JSON-объект для каждого из этих эндпоинтов. Различаются только текстовые значения в полях **tier** и **name**.

**Расшифровка полей верхнего уровня**
- tier (string): Ранг лиги. 
- leagueId (string): Уникальный UUID-идентификатор конкретной лиги в системе Riot Games.
- queue (string): Игровой режим. Самые частые значения: RANKED_SOLO_5x5 (одиночная/парная очередь) или RANKED_FLEX_SR (гибкая очередь).
- name (string): Внутреннее лорное название лиги, генерируемое игрой.  
- entries (list): Массив объектов с игроками, отсортированный хаотично.

**Структура объектов внутри entries**
- puuid (string): ID игрока,
- leaguePoints (int): Текущее количество очков в лиге (LP). 
- rank (string): Внутритировый дивизион. Для Challenger всегда равен "I".
- wins (int): Количество побед игрока в текущем сезоне.
- losses (int): Количество поражений в текущем сезоне.
- hotStreak (boolean): Флаг серии побед. true, если игрок победил в 3 или более играх подряд.
- veteran (boolean): Находится ли игрок в этой лиге более 100 игр.
- freshBlood (boolean): Новичок лиги. Показывает, попал ли игрок в Challenger совсем недавно.
- inactive (boolean): Флаг неактивности (аккаунт теряет очки за пропуск игр). 

In [11]:
# Список, куда мы будем складывать датафреймы по каждой лиге
all_dfs = []

In [12]:
# Полностью очищаем старые обработчики, чтобы избежать конфликтов в Jupyter
#for handler in logging.root.handlers[:]:
#    logging.root.removeHandler(handler)

In [13]:
log_file = base_dir /  'all_players_data.log'
logging.basicConfig(
    filename=log_file,
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    force=True,  #Сбрасывает настройки Jupyter
)

In [14]:
def gathering_of_league_players(region, league_name, queue, my_headers, all_dfs_list):
    """
    Используя API /lol/league/v4/{league_name}leagues/by-queue/{queue}
    собираем puuid игроков региона region и лиги league_name.
    Все локальные переменные стираются из памяти на выходе.
    """
    print(f"Запрос списка игроков из лиги: {league_name.upper()}...")
    logging.info(f"Запрос списка игроков из лиги: {league_name.upper()}...")
    
    # Формируем URL в зависимости от лиги
    # Эндпоинты: /challengerleagues/, /grandmasterleagues/, /masterleagues/
        
    #url = f"https://{region}://{league}leagues/by-queue/{QUEUE}"
    url = f"https://{region}.api.riotgames.com/lol/league/v4/{league_name}leagues/by-queue/{queue}"
    print("Итоговый URL:", url)

    time.sleep(1.6)  # Пауза перед каждым запросом - максимум 100 запросов за 120 секунд.
        
    try:
        response = requests.get(url, headers=my_headers, timeout=5) # запрос
        
        # Если ключ устарел (403) или лимит исчерпан (429), этот блок защитит от вылета
        if not response.ok:
            print(f"❌ Сбой запроса {region.upper()} {league_name}: Код {response.status_code}")
            logging.error(f"❌ Сбой запроса {region.upper()} {league_name}: Код {response.status_code}")
            print(f"Ответ сервера: {response.text}")
            logging.error(f"Ответ сервера: {response.text}")
            return

        data = response.json()
        
         # Полезная информация о лиге
        print(f"\n Успешно получено!")
        logging.info("Успешно получено!")
        print(f"Лига: {data.get("name")}")
        logging.info(f"Лига: {data.get("name")}")
        print("Тир:",    data.get("tier"))
        logging.info(f"Тир: {data.get("tier")}")
        
        # 'entries' — список игроков в этой лиге
        # Каждый элемент — словарь с данными об одном игроке
        entries = data.get("entries", [])
        players = data["entries"]

        print("Количество игроков:", len(players))
        logging.info(f"Количество игроков: {len(players)}")
        
        league_matches = set() # Сюда собираем уникальные ID матчей за май

        # Создаём временный DataFrame для текущей лиги и региона
        current_df = pd.DataFrame(entries)

        # добавляем столбцы регион и лига
        current_df['league_type'] = league_name
        current_df['region'] = region

        # Добавляем текущий датафрейм в общий список
        all_dfs.append(current_df)
        print(f"Добавлено игроков: {len(current_df)}")
        logging.info(f"Добавлено игроков: {len(current_df)}")
            
    except requests.exceptions.Timeout:
        print(f"❌ Превышено время ожидания сервера для {region}!")
        logging.error(f"❌ Превышено время ожидания сервера для {region}!")
        return    
    except requests.exceptions.RequestException as e:
        print(f"Ошибка соединения с {region}: {e}")
        logging.error(f"Ошибка соединения с {region}: {e}")
        return
        


In [15]:
# Идем по регионам
logging.info("Скрипт запущен")

for region in array_regions:
    route = get_route(region)
    print(f"\n=== Начало сбора данных для региона: {region.upper()} ===")
    logging.info(f"\n=== Начало сбора данных для региона: {region.upper()} ===")
    
    # Идем по лигам
    for league in array_leagues:
        gathering_of_league_players(region, league, QUEUE, HEADERS, all_dfs)
        
        gc.collect() 
        
        # Задержка для соблюдения Rate Limit
        time.sleep(2)
        
# Проверяем, удалось ли собрать хоть какие-то данные
if all_dfs:
    # Объединяем все маленькие таблицы в один огромный DataFrame
    final_df = pd.concat(all_dfs, ignore_index=True)

    # Сортируем игроков по очкам внутри их региона и лиги
    if "leaguePoints" in final_df.columns:
        final_df = final_df.sort_values(by=["region", "league_type", "leaguePoints"], 
                                         ascending=[True, True, False]).reset_index(drop=True)

    print("\n=========================================")
    print(f"Сбор завершен! Общий размер таблицы: {final_df.shape}")
    logging.info(f"Сбор завершен! Общий размер таблицы: {final_df.shape}")
    print("=========================================")
    
    # Сохраняем один общий файл на диск
    output_file = os.path.join(base_dir, "all_players_data.csv") if base_dir else "all_players_data.csv"
    final_df.to_csv(output_file, index=False)
    print(f"Все данные успешно сохранены в файл: {output_file}")
    logging.info(f"Все данные успешно сохранены в файл: {output_file}")
else:
    print("\n❌ Не удалось собрать данные. Общая таблица пуста.")       
    logging.error("\n❌ Не удалось собрать данные. Общая таблица пуста.")


=== Начало сбора данных для региона: EUW1 ===
Запрос списка игроков из лиги: CHALLENGER...
Итоговый URL: https://euw1.api.riotgames.com/lol/league/v4/challengerleagues/by-queue/RANKED_SOLO_5x5

 Успешно получено!
Лига: None
Тир: CHALLENGER
Количество игроков: 317
Добавлено игроков: 317
Запрос списка игроков из лиги: GRANDMASTER...
Итоговый URL: https://euw1.api.riotgames.com/lol/league/v4/grandmasterleagues/by-queue/RANKED_SOLO_5x5

 Успешно получено!
Лига: None
Тир: GRANDMASTER
Количество игроков: 800
Добавлено игроков: 800
Запрос списка игроков из лиги: MASTER...
Итоговый URL: https://euw1.api.riotgames.com/lol/league/v4/masterleagues/by-queue/RANKED_SOLO_5x5

 Успешно получено!
Лига: None
Тир: MASTER
Количество игроков: 10000
Добавлено игроков: 10000

=== Начало сбора данных для региона: NA1 ===
Запрос списка игроков из лиги: CHALLENGER...
Итоговый URL: https://na1.api.riotgames.com/lol/league/v4/challengerleagues/by-queue/RANKED_SOLO_5x5

 Успешно получено!
Лига: None
Тир: CHALLEN

In [16]:
logger = logging.getLogger()

# Закрываем и удаляем все обработчики, которые держат файлы
for handler in logger.handlers[:]:
    handler.close()  # Закрывает сам файл внутри операционной системы
    logger.removeHandler(handler)  # Отвязывает его от логгера

Данные собирались в период 15.06.2026. На количество и качество данных влияли следующие факторы:
1. Сброс рангов и старт Нового Сезона. 28 апреля 2026 года официально завершился 1-й сезон («For Demacia») и начался 2-й сезон. В апреле 2006 был конец сезона и база игроков в лигах Challenger, Grandmaster, master была максимально заполнена. В начале июня 2026 - прошел всего месяц со старта нового сезона. Лиги Мастеров и Претендентов еще не успели заполниться людьми физически.
 
3. Квоты и ограничения на места в топе: Challenger и Grandmaster имеют жесткий лимит на количество мест (например, ровно 300 и 700 мест на регион). Но ранг Master лимита не имеет — туда попадают все, кто набрал нужное количество LP. В конце сезона в Мастерах находятся десятки тысяч игроков. В первый месяц сезона там находится лишь небольшая группа самых активных «киберспортсменов» и стримеров.
  
3. Ограничение пагинации. Поскольку количество игроков уменьшилось именно в самом сервере Riot, метод data.get("entries") выдает чистый, актуальный на сегодня список игроков, которые успели зайти в топ-лиги к маю 2026 года.

Ближе к июлю-августу 2026 года (к концу 2-го сезона) этот же скрипт снова начнет собирать по 45–50 тысяч игроков, так как Мастер-лига опять разрастется.

## Запрос информации о рейтинговых матчах за текущий месяц у 50 активных игроков по каждой лиге и региону

**Рейтинговым (ранговым)** считается матч, исход которого напрямую влияет на ранг игрока и его очки лиги (LP). В League of Legends есть три основных рейтинговых режима. 
Матч является рейтинговым, если его **queueId** равен одному из следующих чисел:
- 420 — Ranked Solo/Duo (Одиночная/Парная очередь на карте Ущелье призывателей). Самый популярный соревновательный режим.
- 440 — Ranked Flex (Гибкая очередь для команд на карте Ущелье призывателей).
- 1160 — Ranked 5v5 (Возвращенный командный режим для готовых групп из 5 человек).

Используем эндпоинт **GET /lol/match/v5/matches/by-puuid/{puuid}/ids**, который возвращает список матчей конкретного игрока (по puuid), в зависимости от переданных параметров.

Параметры:
- startTime: Epoch Timestamp начало периода
- endTime:  конец периода
- start - Начальный индекс (По умолчанию 0)
- count - Количество матчей (принимат значения от 0 до 100, по умолчанию 20).

**Для увеличения скорости сбора** ограничим количество игроков, для которых будем собирать информацию. 
На предыдущем шаге мы собрали данные 47 тысяч игроков по трем лигам и двум регионам. Теперь **отберем по 50 самых лучших и активных для каждой связки «регион + лига»** по двум ключевым показателям: очки LP (качество игры) и количество побед/матчей (активность). В Riot API за это отвечают поля leaguePoints (очки), wins (победы) и losses (поражения).

In [26]:
file_name = base_dir /  'all_players_data.csv'

In [27]:
# Т.к. скрипт запускается частями, то открываем полученный на предыдущем шаге датафрейм final_df с PUUID игроков, чтобы не собирать данные вторично
final_df = pd.read_csv(file_name, encoding='utf-8', encoding_errors='ignore')
display(f"Файл загружен успешно: {file_name} | Строк: {final_df.shape[0]}")

'Файл загружен успешно: d:\\datasets\\LOL\\all_players_data.csv | Строк: 21631'

In [28]:
# Создаем столбец общей активности (суммируем победы и поражения), если эти поля есть
if 'wins' in final_df.columns and 'losses' in final_df.columns:
    final_df['total_games'] = final_df['wins'] + final_df['losses']
else:
    # Если информации о поражениях нет, берем только победы как маркер активности
    final_df['total_games'] = final_df.get('wins', 0)

# Сортируем ВСЮ таблицу по приоритету:
# Сначала идут те, у кого больше всего очков LP (leaguePoints).
# При равных очках выше будут те, у кого сыграно больше матчей (total_games).
final_df_sorted = final_df.sort_values(
    by=['region', 'league_type', 'leaguePoints', 'total_games'],
    ascending=[True, True, False, False]
)

# Группируем по региону и лиге и забираем строго ТОП-50 из каждой группы
df_top = final_df_sorted.groupby(['region', 'league_type']).head(50).reset_index(drop=True)

# Проверяем результат
display("=== Статистика после фильтрации ===")
display(df_top.groupby(['region', 'league_type']).size())
display(f"\nОбщий размер новой таблицы: {df_top.shape[0]} игроков.")

# Перезаписываем переменную для скрипта матчей
limited_df = df_top

'=== Статистика после фильтрации ==='

region  league_type
euw1    challenger     50
        grandmaster    50
        master         50
na1     challenger     50
        grandmaster    50
        master         50
dtype: int64

'\nОбщий размер новой таблицы: 300 игроков.'

In [29]:
# Сохраним ТОП-50 игроков по каждой лдиге и региону в отдельный файл
file_name = base_dir /  'top_players.csv'
df_top.to_csv(file_name, mode='w', index=False, header=True, encoding="utf-8")

In [30]:
# Временной интервал - май 2006 
# Начало
start_of_month = datetime(2006, 5, 1, 0, 0, 0)
start_timestamp  = 1777592400  # 1 мая 2026 00:00:00
# Конец 
end_of_month = datetime(2006, 5, 31, 23, 59, 59)
end_timestamp = 1780270799    # 31 мая 2026 23:59:59

display(f"Старт: {start_of_month} -> {start_timestamp}")
display(f"Конец: {end_of_month} -> {end_timestamp}")

'Старт: 2006-05-01 00:00:00 -> 1777592400'

'Конец: 2006-05-31 23:59:59 -> 1780270799'

In [31]:
# Список PUUID  игроков
player_puuids = limited_df['puuid'] 
unique_matches = set()

Чтобы скрипт не потерял данные при возможном сбое сети или ошибке API, будем сохранять промежуточные итоги по каждому игроку в один общий датасет сразу после того, как закончим скачивание его истории. И далее будем сразу записывать этот обновленный датасет на диск (в CSV), реализуя полноценную систему чекпоинтов.

In [32]:
# Выходной файл
output_file_csv = base_dir /  'players_match_month.csv'

In [33]:
def count_lines(filename):
    """ Функция считаем количество строк в файле, не открывая его.
    Необходимо для минимизации расходуемой памяти
    """
    with open(filename, 'r', encoding='utf-8', errors='ignore') as f:
        # Считаем перенос строки для каждой записи, не загружая текст в ОЗУ
        return sum(1 for _ in f)

In [34]:
# Проверяем существует ли уже файл на диске: если да, продолжим сбор с того места, где произошла остановка, если нет — создадим новый общий датасет.
if os.path.exists(output_file_csv):
    main_df = pd.read_csv(output_file_csv)
    # Собираем уже обработанные puuid, чтобы не запрашивать их повторно
    already_processed_players = set(main_df['puuid'].unique())
    output_rows = main_df.to_dict('records')
    print(f"Найдена существующая база данных. Игроков уже обработано: {len(already_processed_players)} players.")
else:
    already_processed_players = set() 
    with open(output_file_csv, "w", encoding="utf-8") as f:
        f.write("region,league_type,puuid,match_id,search_month,collected_at\n")  # Заголовки столбцов таблицы
        print(f"Файл {output_file_csv} не найден. Начинаем заново.")

Файл d:\datasets\LOL\players_match_month.csv не найден. Начинаем заново.


In [35]:
def collection_of_matches_for_the_month(puuid, player_index, league_type, region, my_headers):
    """
    Функция скачивает матчи для ОДНОГО игрока и сразу дописывает их в CSV на диск.
    Память полностью очищается после завершения работы функции.
    """
    display(f" Начинаем сбор для игрока: {player_index}: {puuid}")
        
    route = get_route(region) 
    
    # Запрашиваем сразу до 100 матчей за один запрос
    url = f"https://{route}.api.riotgames.com/lol/match/v5/matches/by-puuid/{puuid}/ids"
       
    params = {
            "startTime": start_timestamp,
            "endTime": end_timestamp,
            "count": 100
            # в MATCH-V5 есть баг: совместное использование фильтров времени (startTime) и queue часто возвращает пустой список [].
            #"queue": 420,  Только Ranked Solo 
        }
        
    try:
        response = requests.get(url, headers=my_headers, params=params, timeout=5)
        
        if response.status_code == 429:
            logging.info("⚠️ Лимит запросов! Ждем 10 секунд...")
            time.sleep(10)
            return
            
        if not response.ok:
            logging.info(f"❌ Ошибка API для игрока {player_index}: {response.status_code}")
            return

        match_ids = response.json()  # Получили реальный список ID матчей от сервера
        
        if match_ids:
            temp_rows = []
            current_time_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            
            # Фиксируем месяц поиска (например, "May 2026") на основе переданного timestamp
            search_month_str = datetime.fromtimestamp(start_timestamp).strftime("%B %Y")

            for m_id in match_ids:
                temp_rows.append({
                    "region": region, 
                    "league_type": league_type,
                    "puuid": puuid,
                    "match_id": m_id,
                    "search_month": search_month_str,
                    "collected_at": current_time_str
                })
            
            # Создаем временный датафрейм только для ОДНОГО этого игрока
            player_df = pd.DataFrame(temp_rows)
            
            # Проверяем, существует ли файл. Если нет — пишем с заголовками, если да — просто дописываем строки
            file_exists = os.path.exists(output_file_csv)
            
            # Сохраняем на диск методом дозаписи (mode='a'), не нагружая оперативную память
            player_df.to_csv(output_file_csv, mode='a', index=False, header=not file_exists, encoding="utf-8")
            
            logging.info(f" Игрок {player_index}: Успешно сохранено {len(player_df)} матчей на диск.")
            
            # --- ПРИНУДИТЕЛЬНАЯ ОЧИСТКА ПАМЯТИ ---
            del temp_rows
            del player_df
            gc.collect()  # Сборщик мусора физически освобождает ОЗУ [gc]
               
        else:
            logging.info(f" Игрок {player_index}: Нет данных за период.")
            
    except requests.exceptions.Timeout:
        logging.info(f"❌ Превышено время ожидания для игрока {player_index}")
    except Exception as e:
        logging.info(f"❌ Сбой на игроке {player_index}: {e}")

In [36]:
log_file = base_dir /  'players_match_month.log'
logging.basicConfig(
    filename=log_file,
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    force=True,  #Сбрасывает настройки Jupyter
)

In [37]:
display(f"Начинаем сбор матчей за период {start_of_month} -{end_of_month}")
logging.info(f"Начинаем сбор матчей за период {start_of_month} -{end_of_month}")

# По игрокам
for player_index, row in limited_df.iterrows():
    puuid = row['puuid']
    # Если игрок уже был успешно сохранен ранее — пропускаем его
    if puuid in already_processed_players:
        logging.info(f" Игрок {puuid[:15]}... уже обработан ранее. Пропуск.")
        continue
    league_type = row['league_type']
    region = row['region']
    collection_of_matches_for_the_month(puuid, player_index, league_type, region, HEADERS)
    
    # ПРИНУДИТЕЛЬНАЯ ОЧИСТКА ПАМЯТИ
    # Стираем временные связи текущей итерации и очищаем ОЗУ
    gc.collect() 
    
    # Соблюдаем лимит Riot API (1.2 - 1.5 секунды)
    time.sleep(1.4)
logging.info(f"Завершено. Лог файл - players_match_month.log")

'Начинаем сбор матчей за период 2006-05-01 00:00:00 -2006-05-31 23:59:59'

' Начинаем сбор для игрока: 0: 7qDgswqpOmHpVyAQO-085TwvVSp5kuojnI7ehOzZiBQX4JksjCF3bVeM-9UyMQXuYNJbX-Gn3gKeog'

' Начинаем сбор для игрока: 1: D8z6JW2Ea_FfL6zRR98wcFmaMpK1KZRbYBGS3pIySYqNC06W4ai6nN3U_O2iv9ML6IS01wsYsIpI7w'

' Начинаем сбор для игрока: 2: Sb7usV3fHduwHxZoXjwHdaVxF1T2LhaBHejvJdqOvRhZv_pL3fWVrgjwh9DDpM-oS-LBVGFhrn4y0g'

' Начинаем сбор для игрока: 3: ZIemTzkIPML6oScobsLKDZZCmGXzGA9yHbJAymoaqRNFopIfvWFpdcrc38PNm1p0XFOI67dvaqnzRw'

' Начинаем сбор для игрока: 4: tlYvTzzoBLVbP6ZISTMT9V8mfibz4nNJVHkWbMSK5xlzF6mz3bE57OEYGuit3KdJS9AiBMSdE9aaiw'

' Начинаем сбор для игрока: 5: prRdc41QhYfj83Zi4D_wd4NRcAaTLHeWb-MyKxYd2ILBnu4p5qcPMyEJkbAZ9k_hnKdbEvjBuZoOsQ'

' Начинаем сбор для игрока: 6: IdYl1YKXllJR3VeiApjjlqo558cORhOBeRnT-samjUytgAoN1foWEcnC-EMz4RDr8Mm5v6pGB_mOUg'

' Начинаем сбор для игрока: 7: bfgHxP9RW7YvYdxR_-s9FuNPIHZOPeI7aAQXiRSAgclRepAZVGjRZksiGDPto8wooTxkRIACKgcn_w'

' Начинаем сбор для игрока: 8: VPMqFtCbfZ2N8LJSACo2MlWbGyi7HJtSMslvJXpczPNcinp01CRf8NwPY7Zlo0RnynJV9kI1BSPC9Q'

' Начинаем сбор для игрока: 9: AszXvP1J7v4AuAbpXQEtXG_UhevfhOfm_dt2p11wxMFJ8OZK5q6j_LCMJF_m7_o8apphqftaJpXfcQ'

' Начинаем сбор для игрока: 10: 8YnyL131MdgNpBHqIrlXpEjEhE5GksgoHifQjoDZgb1I9N0NvlPPizt8A94Hx9lC7krqKfhte5MDfw'

' Начинаем сбор для игрока: 11: vy7PPh2OzzRnKxfaWiTkuwx7yqeTECaIQXAumEnzKgXEbZnr_ODT-oNozEGkT1GKNfblCeJog1ZX6g'

' Начинаем сбор для игрока: 12: s0P1jL0P0UC3xWQd138CHi27r4q5dSUmdGPu6oCnXpvBnrBpxpqRgosNSibDGLInaGgkoS3Xg_GpeQ'

' Начинаем сбор для игрока: 13: 3M6llYE1Noh1gewb2OaBhLFrYMBb7q2OD0VT1xd64Ja1guw_dCgm0AbNQolv4a99kD9g8JUkkG7Mzw'

' Начинаем сбор для игрока: 14: xqFPZPXPvjicRmx6ilEuajWzIjbm7f0b08ebS5GROF8TrPjwz1EjUtthXy48sh8423T3nZeQYzxJSA'

' Начинаем сбор для игрока: 15: FrMN4u_cEW2NET8CjU15wh2Aggb-eXTatrfTvyXjxlPxBgIk0CrMxoBo2I4a0VXsfh9b0S99TJoaiQ'

' Начинаем сбор для игрока: 16: vn-n_iYiv9Q-pkID_V6lTIxEm_s7VB0AWUgK4gzq79u3fiJDuMqE_vSvrMhiiuf-qTQPTCwi3HjGIw'

' Начинаем сбор для игрока: 17: PDB2fhwo-mgy_WAlFFIl4uRaCbQyvwB-IupTQ7GALKPrXg2cAvpmhdFjUSQW-iXyiLVN2on_5z_sXA'

' Начинаем сбор для игрока: 18: -gekwJ3NqiWPawdEzn5mqOADGn3iPkhVWVw4oKMUeujyNMj93XM-tqwpZHyabs1l-7J66x8VmX1A9Q'

' Начинаем сбор для игрока: 19: FSGoTlY72Tq9xxlOGQ8Ydx-ZsFZ2G-eTpj9K03rBROM_CGD7CuCVkyqp9IqXkEx2KxZoR3bOsH2neg'

' Начинаем сбор для игрока: 20: cNaPab73aRrGjs3vahpTadVU5OoLj4yq2Oeg6tweZCpIlR21GAgsia1jwpBECiZ4MfuB_mXDP48w6g'

' Начинаем сбор для игрока: 21: 37PxkYvrNZZo1idh_3kL9OGdAQSa-vvWib3VzYklAbhNvfmYr0ADPMLwuidV-udYdvsXax6S26YA0g'

' Начинаем сбор для игрока: 22: Yzqa04GguxflhWmkQTnQx6HNAKtP3aW52DHZVqJZEy1VVJuMA83wPBB-kEgBvxbs4pPro4-dEi2qWQ'

' Начинаем сбор для игрока: 23: _4prOVjRzPZk9tF0aKTOZrPbBH7yfDQ0VT6q2YXTHj7KhnnggA5tBBlcUkKuA671um3-k6Mk7CovrA'

' Начинаем сбор для игрока: 24: 7_mML5dymsEFh0JEhQt8zv_Iti0fAKBpXOFV3lwhAKVAer02FFKeeBQpdO97mD73Vk7ekyWMbEAiYg'

' Начинаем сбор для игрока: 25: bG2R4sUly7wH1rtObqtrbXAK5vO80NwSDflykO7zppyTpiXoFBIYpfGezSa0jfovgN21dRvPqCd-Wg'

' Начинаем сбор для игрока: 26: 5C2uskEiO6bvwGClRHTC5uuxV5EXlJ3MTnBKEOfGqrRp6IhiGdBrRbp5uO6g7AdYylaaHLK_Y7OZTA'

' Начинаем сбор для игрока: 27: gRCHBbKgZIgVJEnXVge8rt5NPNHNn_fQbglvihvuMnwbb4vYCDIysSPEjmyM_wAjJKqRIVUDeCd-AQ'

' Начинаем сбор для игрока: 28: 0Eirt0Bax9fkuMSgNxHi-zqPVKBXsCXE_aHcNv6j1FEg7xdIEgWQ4dVY6dfawhvNUY8a0MrDRqyi0g'

' Начинаем сбор для игрока: 29: 7aX4K6d3mnD23ZlRVHBWna-K3ojysLm8IHQ2gbWDwLPG9UEZ5_2VYpicqG6sIvy9XJD4suoUk46aXA'

' Начинаем сбор для игрока: 30: wN8LNRSeB558fGZ5BdMZN01P7LtMjTsmPnVvwsewPW4jCAPZa1oJgNPw87qHW9kBZldsrzmtpSKFtQ'

' Начинаем сбор для игрока: 31: tdzMjHB5vE3bVbUVp5RMLTjra9l58h5laIWMXSWNI4hxRIkOX_sUPPCMEQyClj6SNxAs3zGfHLTOFQ'

' Начинаем сбор для игрока: 32: yVUbHCU2-qRHWTjSoAjBTqhYI24EVueycL0gtzyyuB41UZ4RoKCf5k273wkNq_B5t0MFGiXe_Pjg9Q'

' Начинаем сбор для игрока: 33: PnU5Gx3zRJmioGOgprtukwrK4oiM0bQVBlnCIY3N6EwuuT1slR65vbpWYI8R8nixswUxItd60ujbeA'

' Начинаем сбор для игрока: 34: WSqkPOYTDK_YfdfAzPXfwJd_8fqjzEmWz16s_Xt8Whu1XoM5P5_y8DtdpEMQ8Vs8F2WcaEt6NYwmOw'

' Начинаем сбор для игрока: 35: 1NLp56kKQSzyDNwnB2SDzxMtpSV7tYfXlVchc4UNQSIJ9X-jeQY96Tn5HQRut3MUKBj5VEC1-XOGKw'

' Начинаем сбор для игрока: 36: E8g8VFvcT1Ho0CFxe4-mGKuG8FDmPzjxY3n8fLbJZObBl5PN34L-hos1yB4APMuNs9rzJ4YEqX0eqA'

' Начинаем сбор для игрока: 37: fBRSumcxVZlvttngUeUapsH9nxyyJGk8lPL9TFmzdl4UBRCmPIG4ETILfBdKJSCIP5m5EGyc6sVR4g'

' Начинаем сбор для игрока: 38: E1k4fexF9QZCvCtpBl9zeaT5pOSg9VdqDtHuAidwlvJEhLZKv6qdENev1SLTIFdl2dKunBvNNEzS5w'

' Начинаем сбор для игрока: 39: Bfyb3rrTLeJTenxHlXMbH-GMf6nWekXqpRgkxXpVu3F_335GS6fJGiUi4jGx9hem-wfUEXRnc-F45w'

' Начинаем сбор для игрока: 40: qFe0avHPmkt6TWNyx9ElqVg9T8jAuu_jSi-1DmiO2QtkAEoRYbsoR1tqizfSD1QrmKG8GPUG3d-QiQ'

' Начинаем сбор для игрока: 41: FE89fF6H6pIefI-Ybv94wMMPoYtH79eFm4I8zd8PPHAgn3FC241QIvtzDGp4J5rDXd-CU_yuJuA-Iw'

' Начинаем сбор для игрока: 42: muiBpSU4y9NBHl0dUcuZPM0XrMKjl1xmayG2a-mKJBR1XrLIHg18H7xCAhGE2w0BSaod7OGRX7LF_g'

' Начинаем сбор для игрока: 43: 7pGkxS-p91twI8l0G7V3QbV12QKr10-g78xvXtxM9dUpUDGYNAtpFHE1eSzyRzonBsTMVvY5e9sbFg'

' Начинаем сбор для игрока: 44: ijJOC4iujHugMI5WtAiTxTFuGF0Y0P1RtTOLRseVRqyHx_IGR_KxEd-xidoFH4dH31UOACLOf0mBfQ'

' Начинаем сбор для игрока: 45: 44ZeIxU9FvQ0C9ylRTtmEYMhTHBM20XNjdp0wu5h5jT4UQeowKqAfpv9V3xW2t_Jrf13w1I-d9aB9w'

' Начинаем сбор для игрока: 46: HILl8c8s5OEa8qekImEvtDV4W-p_SF8voPn4xF6Cy8pi6LJLBeyK5QYB7JC6Pl173BK_1Zi55ZaZtA'

' Начинаем сбор для игрока: 47: snWAHC72dKAeMFFAkVc7uzOzf1wP8ovRCZc-G85Bp24kCQS5Uc8hpCnzRrBkqnaZEahqpfoeXEzl3w'

' Начинаем сбор для игрока: 48: eEcW70mmXZK8oW-UUNvT2AGCmrEqrgrv4XR-QMDNzyoJk4qwYyc5G2SVM4DzqUsNQFhzA0FA2j2k9A'

' Начинаем сбор для игрока: 49: YSKn3plEU69n29XbS0N_lbkPz8xKVmPiUGhuhOC2LwWjxs6_ZogJdRhm5u_h0eIDXNwHnGno0nuaIw'

' Начинаем сбор для игрока: 50: izy2h_ssfMljCRg98pGnOyo6BFsAijV-to5t-azcYO4NVswJT0qJc5mErkkrrBDeMlhI_TC3QKBcjw'

' Начинаем сбор для игрока: 51: xxeLEHWX0Vayh-ei3hIexo9t-JFlsRFN7xY2CA5ffPXadYlkqWduKXGyNr9czw9_ONYRiTG4h1TeCg'

' Начинаем сбор для игрока: 52: ZUhoaKandkaRZtLouo8sFVe2OlDK8q52zZsw9JkcoVZ1EZbRkugkcxLA9_4XhMF87brEXJkRAzdLAw'

' Начинаем сбор для игрока: 53: SMzVhFhbVFoQ55si_YJgoIoK_38GyCEopwYEnwctY-2BgqW2YSiY6OOgDFMU8EJRlWjFQiqakAI4vA'

' Начинаем сбор для игрока: 54: Gj1P4f0S7kvEyQuTfNCdzEKbNVZZ2DUuJMcZpkuVFtqGo17IJ_wcJ8yLpyhZ3-ow1LQu6VkRptvpaQ'

' Начинаем сбор для игрока: 55: VZnKJ1J_7LIklGY2MB-COoSwpgbMVXnxQ6qOLAyhCeobRcUZ1OpkVUQSr5QJEZHScKetzubwVk74ww'

' Начинаем сбор для игрока: 56: 9-G6x2hcRtii5U4yItx0c2lloWbO5t7QHgr3SQlLnt3O32BnMz3sxhKdKnVWIUIExo8ocF9rvW1WkQ'

' Начинаем сбор для игрока: 57: uwsYgZvzkXopPvIokTGEpxa98wB3RJJvXO3UHeLkLq4uhdwX6zNGAOvSP19KCLQWzJzazTV_dmy8_A'

' Начинаем сбор для игрока: 58: qvtcccmiN09q7JNiV-dH1ziqZqGxnLG7iyeSO3VDFYDC-tubeCS2ukWe5ZblJQsSeNsPe4vOdfbB_Q'

' Начинаем сбор для игрока: 59: fspAvaLqAYFinlsMfAQWkRKCXAseNwCEp-Gwi_f0ZwCUccChVw8tCaeg4PwVdo-eRcIUMnShIEvZmg'

' Начинаем сбор для игрока: 60: GDLe1xfBxEV_QL9VijBxU0PS_CAIVE4iNpKJ_8Gp7zGZpGZKqQz4nK5fCmmcFDWEGV58361L_o1rVA'

' Начинаем сбор для игрока: 61: F1tSK8a3TnVrGqr9Tx__J2h_Ggbopwf6Uhc3TBXAdJ89IBCt_OvfdJMbBbjZcGDWLDTneycwL75Vww'

' Начинаем сбор для игрока: 62: SmWHrDJZcTwglHwrIjxc9u8q3V6FsTZ7yOynScOafk0gek7SPzwhquvp6apGTbZFKeQm6t0Czk2KNw'

' Начинаем сбор для игрока: 63: FLJx34XsTLLd7Wo_g26_bUkgocrAzHTsM8jlUIDL-HJVyKGyXNKr8JgH4BhD8RNxDoNTKkldMxcp-w'

' Начинаем сбор для игрока: 64: JV9ze0ET3YPqaetTBiQcqoMa4BthiFUmenekxIo56En7PcG2Kvr_e0qtzBXu798SmbTLK3764_SR7w'

' Начинаем сбор для игрока: 65: 7o9vzX148aiRXXHBpgIfCaNBKAsWZd9E85mDBH4-wnxoUaQGThah4s9hmnXRlUT5mZg4L4SBbcNyKA'

' Начинаем сбор для игрока: 66: mdCiAoWGJ_s1yoBkyp215NlnIlhFHoemTOC20pp8bQnKfNP9MAJOqnGHNmQSKmWtuNcOnKkQxDRy5Q'

' Начинаем сбор для игрока: 67: WSJUe_Uoqzrc5GOqIUiARgF7NSbi-vskMbQwX1OVhdd1B_4hzPo4nj_QQxr59OIFmpybQaBwU_FN6A'

' Начинаем сбор для игрока: 68: pjpqrVwPDs3IC9qQ27J-Bo_Jaf0I4RHTGoTSZn5WEgQQCDUi3nOiP6WlFELfQVWofdrMAH1EhDuxug'

' Начинаем сбор для игрока: 69: DwOVJb5SAfMkibyk3jRHO6sIYmiJgJndD9q7vwU69rYJavK4ZB8QbPOfuFRxPUURJrKvvBIbsIfQ8A'

' Начинаем сбор для игрока: 70: V5W_AP3hTzSbHwoy2xwKpbRDJUO7ZzaU1fGKfm-raKyl40U_tQ_h0qrGfkeHO9xrIc7cTZ3g7DFkDg'

' Начинаем сбор для игрока: 71: kz2QTlnjmotsj-C7uem4CH4Q-8L2s5cGlyzPD5-nFOZHq7ISCt7dqoY6abzEC0igj5krKyKtg30cLg'

' Начинаем сбор для игрока: 72: X9EWH9aheyZDjbWrG5TaMAcLWJEfbbckeuN2yvpN2FI07FHAH2ae--Opp3hzb7Povh5js2BiW3SKKw'

' Начинаем сбор для игрока: 73: 3cxGL6L-hk-wyEs24-0m_9KeHWvE4a8JU5Xebp3kauK0jSx58gV0T6dFIiYb3E4jfMGAHH-IpOPyng'

' Начинаем сбор для игрока: 74: 3PySQmTxcJMaWoP8q1sydgnqrk8uNxGtdm8DPw7oHBiUJX0shtK6j9E46kEDPGtEFOEUq-knGMX3lA'

' Начинаем сбор для игрока: 75: RmpLV5okzSjplwRtwrLZnvCXJw1zKf4teXyZpyV073KmgLRbf_DgF6HMRlKo_vJFmipcKNcxcF7rog'

' Начинаем сбор для игрока: 76: EZuxhjCfDHHsdbXSti4rGOaOmOINPS83xHXhnkA9t50Qw4gk0KsgrInzPo_VMak8fMQL14EBV6S7mA'

' Начинаем сбор для игрока: 77: YH21AVnD6AWhz9LFT1SlFIU8aFD7qTIWCUMdzUQCTolicj1qeTNC3nMyjdYfgm8_HFLmvmyruPxDsQ'

' Начинаем сбор для игрока: 78: JwChTkUF-58vEgKQhjHSo8cZV6LBnOQVBhNw7C-Ftr5D4U5LLdNqX5jRPN2yV7G_czYaJLvvRY2KEg'

' Начинаем сбор для игрока: 79: sCujEqTrCqvdb9Tg0b1HFJxH5AvxYMqP0JXxnHsVC6-mBX1t93g7nwDvL3Bsw9fb1oKSUaGs6WcbWQ'

' Начинаем сбор для игрока: 80: a0AgLuM_-qHN6giVoFeS45X1Ten57z-tXigIqxjst0iNcqCJ06N3BCxiodkyBzt7GBV69B-nanI0UQ'

' Начинаем сбор для игрока: 81: ccX0-OGCUEnXKlsriE7IXV91jnAtyrDHwV0_DXeUmjm8JmhGEiJlvnQfqS-k9hwqQC4Szt8Jxr_8Ew'

' Начинаем сбор для игрока: 82: BqUntIR4kcJHFSJoZTRB_8iktdRE9M8wyWE8iZwinh9GyR6Ri0zcN6c690YRp45OPQCytKqZS6s4tg'

' Начинаем сбор для игрока: 83: 3bK1F87kbEvKFjyaiSSh2IELESxSZToG-30UZIEYgJlYUusXAellXsDfoLmBHKeFBp8HvzB-NfRR4g'

' Начинаем сбор для игрока: 84: dhOTl1aIX90t-RrYqo-ICflf3gfRFplYbe7WawjDdFDcgoHPYnstaJLcEVjtE4Izh6-2hsoFUAnP0A'

' Начинаем сбор для игрока: 85: gmBihdKDK8cgciU54PVeghObeHPYlbROd-eQU7v9Ifn4sIadPmb2I0MrhaQ42LCPLoVSKP4xlvb3eg'

' Начинаем сбор для игрока: 86: sWxl7H67SITmBj2qoNVRM8vHL_1yMGQjpEcYb-GtSNnrFwjLAY-2jYd5oESlczz7xB2dR-P2Ac1y-g'

' Начинаем сбор для игрока: 87: W5_-s70r6M1sl-AaKGxNEUVgFmeVhSPa72A45w6ANBbE4MpxBUGyx0UuyMnYpde87qaeThBD-4BZ4A'

' Начинаем сбор для игрока: 88: Wcqhh1aPAuChorOFmp5NxmlbhYpVFo9PI831TIp2I0jqYbVtxyO4g0eJnblAkyxer3wuX8wFVX6TQw'

' Начинаем сбор для игрока: 89: fa6xZjMQCKcSt4WrTc8eR82VbvDqQFY1OwmGo_qiKsagp2tVm9w5Qm43owlQoMcJGaI9mVbAQlAqDA'

' Начинаем сбор для игрока: 90: PqFF7HNyab_5PXbbQQoZpPpoBPHg2Ud155_yRg4WECezHkh2igYwqB74DUsNK9hbwiMbdnIUnn3bDg'

' Начинаем сбор для игрока: 91: HQUSrzDiI6T93-MS1QUMcmGW3Pe8W6BYM6HbQ6WGh06FEPuorhdAvTyNRcBd4AfE53GDCVbiwWy-GQ'

' Начинаем сбор для игрока: 92: OmktZL0Bnor2y6ndUg70clE4JD0PbTNZZaLRz9MOGQPIvAmconkfrANMk7Bsvu5VkGu79BsLnayTPw'

' Начинаем сбор для игрока: 93: 75pk8o40CPKesUv6r3_wQG87L9md-26b92zSyykuRjCaKqB21JincWsXenLGU1B7L787DPXDzH_6jg'

' Начинаем сбор для игрока: 94: soHtu6Sc3NVXAfVP35WDNg1JVxE-nza3Zcc5rO38Ckzh8b_ZKGuk6wpMX9-ypSubc_Y9jSlEMbllHg'

' Начинаем сбор для игрока: 95: XouxUAWdl0fiuTDA_vjBSzLpOCz5C8Nv_xGjsuAZGp4ZT_UO_ZW-WcOU2tZpGPCfY1fNS9nj4gIQmQ'

' Начинаем сбор для игрока: 96: OnuN66iasHJuP4io-AFC5DyK5iucjdAFfG1nNxioVjG1y09rhxKp4WgR_kEEgAC-s-eJQy1EDcRjYg'

' Начинаем сбор для игрока: 97: -zJoO6DdM2uAVAzGOrp5PXbYnPpaVWdh3IeZiiU48yMSg8GKo2QHnZk6TqoNITMX_SDFzW1_M7x1gw'

' Начинаем сбор для игрока: 98: yh4rlhGa7_lKq9PrxhlXY6yZ0-zSdEM36QDII_0wRza2ap79UT1ZlVWCxKSaLSeG4PPvYU2GAZLM5w'

' Начинаем сбор для игрока: 99: oNhe7fxottM3SbZQaXtDoz4SEArxb_ANY7YF2AEKWZFdAvHGRdBCq3QnLgJui8Mxr-5BdWOPE7Pvzg'

' Начинаем сбор для игрока: 100: LycjoZoH2rABzOlg_OhNgVwOhRFO_l0AqVyIqSm_S_HZXyspXcIks7dJr-2xWIqBktpUGb_gS4qlQg'

' Начинаем сбор для игрока: 101: gi1XDZG88vrxGKi937vChOb5cAJtDywXuSvrZZtJFfhxxFlM34p2-oZy9Kwqg_E-VKzlT_Pl41MHLw'

' Начинаем сбор для игрока: 102: bEfwB-fVejlSz7nmyqfq0hY3Abww5UO46EYb7g_fNfN4O2Gd0KPLDoeK2UTmxBVyyEa4-dff9rVLCg'

' Начинаем сбор для игрока: 103: MIMvL70ac4BnVL5x-6cAu1ivx3j5JWf8cNNVGKz9iHnu16ZqlC3wNblljSog7J2s6ZB2VB4HxrjSXg'

' Начинаем сбор для игрока: 104: TgDf_plDJ3PkWwkkhtlKON0UzNmnweKAG8TYBYr7BEt3jps4Vy1mMC6Wfp3rk-QYgzV1QuRYrLY5bw'

' Начинаем сбор для игрока: 105: 8QTQrWeCiEmcNmvI6w8qSE8LfJnjdvRTzQMDhzD_dsj_LSfMuKb0g5jG20ti00CD4mAX_wNjsou01A'

' Начинаем сбор для игрока: 106: Vkhs73BgvcuGTNGzp2mqslRMwlGFt2WEScOVFL8dQRZDfWSuECECNvGJ7-wA4M89UkqpPQgs-pKRXA'

' Начинаем сбор для игрока: 107: FExgwx2mIGy_YV1STTHD33AM7QLh6MiNvwOS5VJsmvp3DxqhhZZaXb09m6KRKGCM9g8-NPToojR9SA'

' Начинаем сбор для игрока: 108: 1V8GzEXjOpOk-GCvI1Nu2rm-y2l2UcJ_CKtXq3AchODcgNb2BvsCsYnWUx2G1SN5xVd8ysEZ2AK9xA'

' Начинаем сбор для игрока: 109: 1mVPcwOje5rzg6a0K2kAJKLQ6pLQsOg7QnkZUAOG9McG7aF8D5loedattiRDvh6n6SMsGP_NeJJslA'

' Начинаем сбор для игрока: 110: IelvKdSFcfaNbGhKXuIUZyNKolyMpbxUp-5EFy1xP7M2wasSSPMfcG9mwYWxFDQPoBLGBzsRkuFDaA'

' Начинаем сбор для игрока: 111: lnWqgvDO044q4JkPNX57rxztkR5reAAYiChyEKU3ucVfzKgUGbJG7wfDjyvzOZZ-8gE6G98fOoGMYg'

' Начинаем сбор для игрока: 112: gKXeA17dq70WVNxW5OtctqHq5WT7FX3xnnwMvtlsSecIMcSSy4SFqpvVQWWVDHcf1BkTxTVRJNwjeg'

' Начинаем сбор для игрока: 113: qTkgYQgt-UADnJgux6usvKl5K3IN_op_xUONTahQJWcHFjQOlcKljYKDNypJHnaub0HWPEKbe4yv2g'

' Начинаем сбор для игрока: 114: N5ei6y7Chv2P9QDgjGx3MHFRdlrvl1PvZjRkt-xAOqSOJFPEPp7hDQBeKAcqgoAXq5FVtc-2PX-SHA'

' Начинаем сбор для игрока: 115: IqtFqUmVyZ-GaY7iw37UrYJCIfNi7lpR7ANRr7rs_XUKDt7KUWfg3HxAfE_oAQfDPqj4rKnCaz7NEA'

' Начинаем сбор для игрока: 116: OyJc4Hvp-peAdCa7fkvSOytaXZAGyoDdfgfSQP-1GfbiueJyM7H1GFaBJPiyDr_ncZfXI9-hUuXoMg'

' Начинаем сбор для игрока: 117: -nXE8GJePlzRAyuxXlhramMzfYBqQuOcS3dSJuDVFX0t3Dzon-IoJ0rupnZG6obyo64GQW7oF96LzQ'

' Начинаем сбор для игрока: 118: uG8UrgU8R5KFFyKEDAMgfLLi5dAGZOFEHSk29n1e1bY0k9SF_5to0nKh9RCJNxRZ6Crv-h_OgR2hHg'

' Начинаем сбор для игрока: 119: LlhP2iu3ZGlEx2OltqN5w-F0skVyeJM6pPIwZqdlg4RIlI2MgkCt_AIlxz2bYcjwGR2bK5qFepyq3w'

' Начинаем сбор для игрока: 120: kl3fLRdvSoQPpFErLx901-Hve9Wn6tNVCMoMYkQ4EcYy5u4HBpyPTiKSuRwyVZCc5xNGp5baJ0ISVg'

' Начинаем сбор для игрока: 121: KXgCpBBLbzu3qUJDgZPELQ61nnQiwfGmiUI5ewGj01tdjIqdB4eMMxO3sg7ltIAGa0zKRC6ftLERHg'

' Начинаем сбор для игрока: 122: Jwk_3NLBvIzHQpSzZjCX897PIQVy8JtEKjrwxwLaBajE43HfyEkQOUASlvnNn8p7G2BJ-VRuSZh6ng'

' Начинаем сбор для игрока: 123: w593tRn1SClbY9LmL66dP5On8JuFyGVvKbt0YO1NPXDTbreNyt9FLMqN29a0EtQWerYNYPaxHgyWUg'

' Начинаем сбор для игрока: 124: Umn54l93XgewTv0JjPZlCIZQoamhd9OgVExYoBTqdJ1dVx1X4bC0Jd9CAra30dqkoQnDobEuwUXO8A'

' Начинаем сбор для игрока: 125: w14f7Oe0IPFGqnp57V7amx8zwp_yFJTRlGWdK3oRB6pPf2TLntwKFhcqBuqNIaBEvUlPoP8b0HfX4Q'

' Начинаем сбор для игрока: 126: 8tInv6MdGnZLqS5b3lXuRPP-LNmQm44N72FmBd8dceJJOQJMLXMulV26lG7_S8axWMVVSRWQesgM5g'

' Начинаем сбор для игрока: 127: MSfB-J8aXKiub5846ZYAyMf7MiVtHUXCaW80aaSKhrmjxYj3K11xolRcJkgleWw1wHIxTdj9e8k19Q'

' Начинаем сбор для игрока: 128: OIQFr1eFoeSZoDnNHz70pLWnPbpyq3mGFOhg36UVHDVpSzd0PAOjNAwmUtvMotoHt_oXOOA0D22jMA'

' Начинаем сбор для игрока: 129: z88NY5AtPye5LCFQXgaVlg4U299_aC5fDMWmcj8-No08Uy18Tckv0wBwtrLlDG0MI8z4uyU2gEeQwg'

' Начинаем сбор для игрока: 130: tDmlD-TEmhqzVtX4SHHAidpcc8l3AYOvxPrgtmdVgqNXfmPdGsefD5tane4fzRaZr1m_DgCZym95_A'

' Начинаем сбор для игрока: 131: wxRMuInpf2iB5tMuE2M53BFcxfY5u_Ew2ZVLiJz1Ck8WrEckxtzyjPH2vAHIrxlhd1tk5bUbDtrC2A'

' Начинаем сбор для игрока: 132: 4SaV9B8_V3WGwlFD4UXo_S5UoDQuGbpGskD5M7VNzHZxGpiWT9RAQHldRaoLoxRXdjTJAzL4jrOayw'

' Начинаем сбор для игрока: 133: zqIZYZydDZIPCFf7E9bIs7zNdHF4Aax4FT68TGHYPFHZzQKJM8ItyQmYJaqPKpZdomro-AvJvvvejw'

' Начинаем сбор для игрока: 134: rqs3Q9EgRaoK7QTXuEYlXIm2tkmps-zi7-XPQTy1kwXUo5zDPqo2rtggo3M4PQ0INUMlVxpPoETI9Q'

' Начинаем сбор для игрока: 135: _SV3XjhwLwsFpmB8qg1V3SEOT05QBv6o5EtDR2FnipbhtoZ1h3BkELSOVZ1BzfWceuv8OfkpXBQr3A'

' Начинаем сбор для игрока: 136: H_oV04AmEuPvAzJtphsP1UGHN3nesN3IUIwvdFbOgy5VIsJhN2dmdA0_lk0QDy7YjSfFYuH6XyYx7Q'

' Начинаем сбор для игрока: 137: ppHJhCOUCVczOA82C1MvzmCI1VAaRI5CJF4yb-z256c6RU5mkxQLjUKxPcOMAJdZyYa7ecXJIFOclg'

' Начинаем сбор для игрока: 138: 6Cku2z814yZ2QqjoDFWLnktrK_rkIFEWCRAwtkQ96LwsUAjpb5s5Y3hZC6Q4XdCI4EdK5eWgFTFDQg'

' Начинаем сбор для игрока: 139: NDvCN-ikmBFbOlFYGFKvGSUWxzsO3V51OEPIHnsEaXLaOwR5_IxVFjp33dPdFZzD9OgZtPuy9piUrA'

' Начинаем сбор для игрока: 140: yZfZxEsOHRl-O2eKljsnvYH9GAoK6Cl21nUGCoK5TKKMQMIZdA4G9kwpuKJ2F5MF6JjSunDQwLC0sg'

' Начинаем сбор для игрока: 141: MQ6gCgFKhgY27odbOZ2qHXSvidhZrC6SwaU0FRehnnOYFh66LHGypPDlYaERU4zRnzl2LvN3aR0GDQ'

' Начинаем сбор для игрока: 142: FO6X5Q4cpIaWggE7oJmKitg1Qh-uiKxyO6HwUs5ZwKyxcRwbNCgXJ4gCI4N5xZMPjlMsV43IpTPIzQ'

' Начинаем сбор для игрока: 143: zAtbDvkZQle-Sar2K-moHeGF3SpjusPwwIcRBf4hcpO4dlROLWCXAfJB7-6zfJSr57yCxLpZbpmXRg'

' Начинаем сбор для игрока: 144: zquZUsshDUYjIHDDvqvYJYYLHggJBeEOw32KVAlD1QgiclUMRl3pudn-9n8k7H0F-8gWtYcx8FpSDw'

' Начинаем сбор для игрока: 145: JpCXGkgryOxc_gpBj3Vi9q41WPJHDfovdRlNULrLhGHZL-EwkxTkONz-bUYysuXwiaGHmUXHXZzPiA'

' Начинаем сбор для игрока: 146: tbEJ405JwUm0rYpCEpiOpsLGG2n3eyecLNzegTYsQfgqSBz09uKEsFwMkAGs6UtnDcMFif5hQdsMsA'

' Начинаем сбор для игрока: 147: xhXl_k9WJrwvZ8fNStbbpRQuUkbfhTLTlB7y-dbqezt9_Cj5GFgenVePcDxjFSiwX1DLNKgtMyapoQ'

' Начинаем сбор для игрока: 148: BAbTZhVoKHK5RzCgDFlVw8B3G2TDcV2RQq-LIBs97Iqv9OmBImAhiU7rmwbYim6CTNiP8bpmETuPCg'

' Начинаем сбор для игрока: 149: 3gE1oIvetiWyHErQl_5L5jevFDL5OPlz2ChXGIlCbq-vqbmheFm4AMIMYQR2jOJsbXwLn4pUSKQqRQ'

' Начинаем сбор для игрока: 150: 7B7rGmNUe_sk_CT5m_C_fiFm8Vu2T3mvzLHD7nqqReGYhHpDUJdPbH3-G4ONfuKh_Wy0PVhEfVYrdg'

' Начинаем сбор для игрока: 151: ZV63rwObkBYHKN8zguuINiNAu7yPj04B_-vWzZF0qeJA5IjXHZNcBb9CRs0kFwm2EhetQOnI0ZdzJQ'

' Начинаем сбор для игрока: 152: eBHjU0Zu1geK6ZgV6yqNVz137YK26LvJ4lH7avex0HHntytQjam8Uw5GcCNHDrfC9aV5ldC7DQ_-CA'

' Начинаем сбор для игрока: 153: 3P5oPveOstmodHehe2G2doXgyvFBSQWLaYjTH275Ez_eBisGA6kNzLJIN0RZw1pHFNw6ALnIQYwMKg'

' Начинаем сбор для игрока: 154: OhBp6uL8ZXGHm38XQebm0nr3DamiK9aXz0S5SSSdAeqjg4UrSJhZbpKs42hW0mfHqHNPEf0DRoXpJw'

' Начинаем сбор для игрока: 155: -JnJv7F0b3OoXT05a9ul6qEJ1FB3bKx-A49fuN5Ve4Jp3uVjKUmLd4jbZfPqu1D8tbF_67J-FxG7kQ'

' Начинаем сбор для игрока: 156: sZksAAOsTAofnB-Iq_jx4S0oixvR4n24mYfACcxFMfWbLH0mrsdXqzPyl2nW0FFoaLMbU694AxF9pQ'

' Начинаем сбор для игрока: 157: vh7uN4on66vxNZk9AJpm4oJKZvXbwxPK_pQEqA40lvUU_cVd63oyTGzqkQcj_FXZTsedu_P3IIKmeg'

' Начинаем сбор для игрока: 158: V3w7c_tyDXV_jWMcSTijb0syvPjTsh1efLqsrfldqqU5PVlDTmfPmzzJBrYhYCtwUqJBkzcV2teHAQ'

' Начинаем сбор для игрока: 159: guK6PU4U7A4ptG0AMU_VGAiis4IJS4Agw1KQ-dBlBCTvbPUouN4hgDZEsCbtuSga_A4muYKLync9nA'

' Начинаем сбор для игрока: 160: z2xJGrLWG6S8qxIbmVVgdTLprA14XIN8W1tq7x2ss9fwvMUmLIglB0yr6oXhPTwqP33iKvH3lL8gaA'

' Начинаем сбор для игрока: 161: RR0BtDTobF-UR5hRw0qOxVEPQ_cKCX_lW85gs_50VyVsHTBAr2ABjSUoUUdriUL_SrONMMJ2cOAm2w'

' Начинаем сбор для игрока: 162: qZWEFCkHW2CyzXMYkgBpcBqc3CmVEihIip5QcRwtKkTNv6--AE0KhVrDwMiZyUAfk7j7dWbwdCrTxQ'

' Начинаем сбор для игрока: 163: nrdCgdHMZNn1Pg-Zfo92qau-Xc4XnADAzpFm4D7b-ygPz1hks98AQDVh2cjHpGb6c5i2Z2o7zU8psA'

' Начинаем сбор для игрока: 164: xJ3sBTGyZ1Pufru5vSjhR_-o0bQTeorYKI9ITY94eomFU70HsixfvglSEf3sgB1xQTbK3NkoOd0sdg'

' Начинаем сбор для игрока: 165: h8Xqu8lTMunkGQz7_sCcHuEvSXl2S7qvCaMtIMdUjt5Blke9xH8HYX94YslC312nhEVNIf_8wP0lzg'

' Начинаем сбор для игрока: 166: yqkmAnLLTnZPdlAXdP0OZR8DIO35GoP0jJs42Q62XJ1IkdX0s3QBWMdU0t2-ekXHcaO9yzhy5NfnIQ'

' Начинаем сбор для игрока: 167: B03TgkL9jGU7koN3jYlHdhBdmyfwywXP9c6H1Lfg0wP7fP914fz5hLk2Zf5DJ477necf2gpfbc8C9g'

' Начинаем сбор для игрока: 168: MBQzxnnZPdBLb0Ej9BPcWh_YTKrVWJtof9w95eJb7FNOpHY3GwoPOpTihdHmAsJxWXM8VlIXpkd0bA'

' Начинаем сбор для игрока: 169: tbYsHJfCai4iI0PVHLwer-I4c4z2ozBxdWtT05u02Kkt3OOKaOkbzW0QiJZnynxV0E7MLn1LxywntQ'

' Начинаем сбор для игрока: 170: OtRFMBfC4V1ExylVmTrMNB4f_R-cADvNYHCyRanOxGWFvxOykm21Y7HY8rL6rDAb5QNedwrdsf8cYw'

' Начинаем сбор для игрока: 171: zejmnT3PnosAxY5v4M5MSbTVKdEnxmiVyXST7F6s-i9eDF1p0DLrivWaaL6nD0ePW6ht1YSjjeF9aA'

' Начинаем сбор для игрока: 172: BorEzE49eLiWyjlkd-aRrKvACY1ByPWZv8D79JXrDDfctfWg5QLQbNX0jHaz1olWoDgsIvgVu7jLiw'

' Начинаем сбор для игрока: 173: 5iNHhX5LC57w55c3JGp1NB5jr0_lF__2yadLXk82zY7zEVSCaNpDKCTtBGEUuFA0Z39_an5INvXO2A'

' Начинаем сбор для игрока: 174: 5lfvmU6H3N7p5Og-Xb-zXAL9OY96YWuXQJWa5l_FI13UP_1HA8O8NtxnV8GMCo0b7v0q0R2Lb51vpQ'

' Начинаем сбор для игрока: 175: ynoiW58smuOB7BSqczed_hokclLzcQBVo3nKy4nSu7at2Q4E8pgAnGUjrdT1Aewom4chUtbmVgKeBg'

' Начинаем сбор для игрока: 176: rHoDamczbPme3CDiKlj7CGAXkKknPLR7A7Gq0LJL1CrmcGBA7a6jL32n0ShgpB5U9vwlRlmfbIupDA'

' Начинаем сбор для игрока: 177: D_goaAzE44LIs6m1-cpEEUyfbEDWl_oPKza23UUNgl7WRKhSR2D4biVdXsevKaPQOfUSVuauKS00Tw'

' Начинаем сбор для игрока: 178: 0X8LbgFq6xDaHaatFQabZFtZvqVxRNI_mNUP347uF7MoCsFnshAzTWpUSjZrFrt4b0m9FuufWt3_AA'

' Начинаем сбор для игрока: 179: Te66hbm1u9hquqZt-KlbYhduNwfT1p3yuBLG826FoSzfGCAEHRmWmCXnomZemH_nwHzbpBCTj2TVKA'

' Начинаем сбор для игрока: 180: uNhTYJ9yEJizA7PmRjmDAI6JCSS08ZuwGE5sW09UbBuHWDI2djllEtU47U-7VdtxLUfJpQJxjh858g'

' Начинаем сбор для игрока: 181: DT2HwZo4SfLZKHlyzRvdKeisf4gynrTN0EHf6AtdncpMtbJmZsSDcMHOeMMeeTeZjRksHUlBUj-GXA'

' Начинаем сбор для игрока: 182: J_5pY0JEYvg5QzDWvWTCVdbhUGlr4bjXf0qqZaeAdD4brhhtGE1cz7iUsGdeSU4sNh8x81dZrCXlCA'

' Начинаем сбор для игрока: 183: rHWigDatUjOgDfUEVg0AEyGzvmhq2ADeC7HIK5SYiIGFoVmrlONsubpxA4eNj592PcONO6-wTDKLsA'

' Начинаем сбор для игрока: 184: -FzyxVCvXgxyf8cwT2KcLepUmhWWrCXMPefwrf0ctpUDS3hM87lUFzV7nEncRy5Rb6p9WwvcpXM3Ug'

' Начинаем сбор для игрока: 185: hsGOfh35F-iA0bgG7om_XUUA2V4-6ju_RYK5UzKMSQ0dnea7ArMZcBaPlCxpIFkOmbg1sIavPZWW9Q'

' Начинаем сбор для игрока: 186: QHEtV8KnS1ACnbX3cWCw9JML4hBsqEIVZuZA0VreL-uEEn7qzWg2NQvRAZl9Q3FG9prMzMD49PeLww'

' Начинаем сбор для игрока: 187: RkbM8nit2rq0QvAQphV1MkvcIUyF2lW7XGIuTbtWFNaSXe33zBT_47FXYmmNqM5k8ChSMdQ_eEmTEg'

' Начинаем сбор для игрока: 188: 4SxEcnAAVs9tQHteZgSFlvY8fLKkvmKYZUQksZ2DyLzcnScpN9gjXJieqxoCGYp7AxNMT7AyH_3g4w'

' Начинаем сбор для игрока: 189: j5YCDC4IdBbSoO7wnDu11S7r-j6LT_yXzRJwN29YccR4o-3bv2GORym9N5aeRboj3fY2EA4wQhoHhw'

' Начинаем сбор для игрока: 190: XuEvKeZc5zdhacR3Z49rnW8ukEllD8C5LxCUSl8VGpoLpO7mFHmIUkBjLXrnC9K7osRlj-Lx4c61-g'

' Начинаем сбор для игрока: 191: Ufnr95GAIbNs0npf4CVX0ITFHxqw4v92H6qOKmhL1LieKoGFRLfMKMOKz7HmbqIj4hKG601ax5unCg'

' Начинаем сбор для игрока: 192: dxbWDCarr2WxCvodamvblVPi9XHFqCYBb31ZfDaL8WSCZ54t_l8zpeXSOaigKkOMYJxMPxm0cSvnHQ'

' Начинаем сбор для игрока: 193: bU6x8goFbqSuKpu5QumRztap3wpcnNdWNEpfVJhxvLmPh1Bw6uEb_jYYiDjB2zjS0603kI-h9yE19Q'

' Начинаем сбор для игрока: 194: EGA9bxSj-A3JK-cp6miVj-1F3xtBzGbYBZyj3eL2R1xMAZNt8U3MYKSwtH3hwV7IBJAzMppQPmEzjw'

' Начинаем сбор для игрока: 195: drJpDlub-INZYjwPjT6GfuTiw0aaFkc6-xn2x8wcY3Abh1tDQ4Xi-EhT1F2B_3OZPT5XNMys3LNatg'

' Начинаем сбор для игрока: 196: tM_5V3233LnM7-yUChamCND8O2v9qeRYVTbZHM7squyyIic6JEE_652qB2dtHfen25NIf43YzNKQOw'

' Начинаем сбор для игрока: 197: 1dQuiBx8tcSvvaI8_x5oePw9KRqLr5Nxgf2OzmahYyMK9wsZ-O6EGKRT-VIChOT224XSGrQkTnvZVg'

' Начинаем сбор для игрока: 198: gCKoQmNWZXp_ndr5-2jFHykb2D-Xk_ufDTiZ6gim01Y2OSLksDTjAqsz95VIb5mQXdlIyqOL2vZKBQ'

' Начинаем сбор для игрока: 199: -0jl2WT53Aq-GXDNBY_QU74MTjRYGvEN3ktuxskQ5jeVbSMWi6GeUAN_aoGmt56Ulvy9OnBaWdQKKA'

' Начинаем сбор для игрока: 200: YObdAhes1nPnMSHy1Fd41aUPyHCkFFXqCL1G3gTNf3xUuO1SFY9lS9Tof0XJBI5lIoVs5tIvnGhxZA'

' Начинаем сбор для игрока: 201: Smtow_tyEpZs_BjWw9s37075LbEk2NUj6nKdu5ZjpkRqLSfz-oLw4yp6ysN97x5JSAbDQPP_r7lztg'

' Начинаем сбор для игрока: 202: YuSDwpHLsEtB7tJYUdU37Sxsp7Nl3kk7Ts5Zne3NDOyQX7BY9ptPNZ-XBUi_sdNytif0vIRVd13IYQ'

' Начинаем сбор для игрока: 203: rYRxjoQEs1WYMf2u-aMMFZlr4i49RoRxeHzh8KpvR_AixuFYMwVcHHFvElcD_68GG0BTloEke-yYfQ'

' Начинаем сбор для игрока: 204: z_oqQkpwIrItw9-eJtXxYamJtbiYjldBooR19IV0sNnCam6jRh_8sVmlCBp3_MVPiqt-5qDFcDY4Dw'

' Начинаем сбор для игрока: 205: kpG0BaWzcZpkQt5wxvrugWe9YXkKrrDig8Tby5mc9Db6LDeWkEYI9iMVMO5SGc81bHHey3UqHu5GJw'

' Начинаем сбор для игрока: 206: 2HBT8Gm5i9nThxPovRLVchbwOF_lGxTOYFx9ToZfyISGAVYy6Ce-klOqb7S-31mxzwZNQxmkYezLpQ'

' Начинаем сбор для игрока: 207: HAiQzNc8RrfgcO2cmLPoNl5hDdHzzyTXTUY7nZXl6pNYLAstJmgUh1FP07lZuV6qt8dlfU_TwLNdbQ'

' Начинаем сбор для игрока: 208: LUx6NhWe6eNzhLUAjvauKm7rLHltFejJnX3LtkNZX1GQJGiyWiZSeeB56ZqHEtOlURAYoKyY8AWTiw'

' Начинаем сбор для игрока: 209: 85c0CNRK4VK8PbyXcoNLlNu7AqiQJ9K1RaErSNMClx6lBLjTv1jP4PK8mIkuwQGs32rYREry-6lo9g'

' Начинаем сбор для игрока: 210: Q5ExGBk_cYSa2uDaShMkjpIQfQ4YdCb3siBk7yQzxwuaOos4W3_dxzwFCoH-c9Gqv291w4e3Uwzvqw'

' Начинаем сбор для игрока: 211: OxR6CUeYztbcaUPKai_LjF_l5CG9_NKEVCvORtdrrJRcm4pcVZ9L_weE-QJ93ABbddXnpHQNwMdtPg'

' Начинаем сбор для игрока: 212: kqa2fPvzg_0Cm7728FL_UbGaSEbVG0QH5-k6dZU6O-n8CAh5qpnMmWLdCi7A2hsTsfp4ZRrd1s-1Zg'

' Начинаем сбор для игрока: 213: dFwwcRsaVNfIgjZ9cmH2iZiJZRb59bly3lHOl0YlSbPdciGkWfektXsnECWdoLSKDvEaW_0ZjKPlOA'

' Начинаем сбор для игрока: 214: JOAA3YEz4qPCLYyHS-qCkXVl7HFRzV6aZu_g4EOwoXBLHnDdmFe6ZwLiRf4Eiul5o5F_NK5cz7Aucw'

' Начинаем сбор для игрока: 215: 0bc2nGPME6jxoA5E7Fz8zRHekMm82vRTbPJzwwK-NjqUHDZy3_mxdtvAgf7Z4pmuIz8jXU3MkfREqg'

' Начинаем сбор для игрока: 216: YGGMRtXp3hZ8JR1lBdqf8FeQH2qkpRSXz18PzBMVxK5yrDMhcylVqkcak1RgzH9xS5GG7Vljwg2i0w'

' Начинаем сбор для игрока: 217: 6cOFwLEPIjfHbrPCiB-SYUZjLIbGNjrqvZNfBuxk4ZSJmq97atvRBAKiyQ8a7HPJIKrFsyGb381D1w'

' Начинаем сбор для игрока: 218: l438aXKn6-BpCpn7QWj2ejeJJ13nRJgwTyv3Gn811hhVQUKXXZuww_WO5AD_VouVB_n5E2NHaPtbyw'

' Начинаем сбор для игрока: 219: bBv2rQRtBJgugnQ9IhG4GGJuYc8b7dfFKuwdHeaIcjZ-3I06KNbMvkh7fQBJKIw2pBp91PnVP4fSog'

' Начинаем сбор для игрока: 220: zF9K5RAfgY7gMfuiawsxc7Xn9tYRS9egafjggBoiDW7wVNeCf3eqkjBRYgI4uIe_o0YKy2pUwEY1ZA'

' Начинаем сбор для игрока: 221: fWyEEORxHDXWSJpEjHOVUvd8H6EEZLO2pPrkwZJZ-jn36k4NvPD8hQ3kJlkaUamIHVKNJyy9yH3YQg'

' Начинаем сбор для игрока: 222: EXA6s6CTAFw3gj601KsG9x-IDQvBTonbVYdGqYOfahTpUcA9HqexWF-Gc2AHhgQMtrD2IjBC5bP-kg'

' Начинаем сбор для игрока: 223: fe8CSseF4WRA_eBHABldhuzL_teajA7IQU3bx2-dDKziI_xxNVcgA6epvZMKHwLUD98YAIsMVsXKTw'

' Начинаем сбор для игрока: 224: uEyhR6Qv70Ml_TIFjol7mVSdFs7ojKbKKh6chJB8jylWQ1alWEAC7vTQlo_N8S9rcyTvAbk9H_b2Gw'

' Начинаем сбор для игрока: 225: XhzZV-JwVtjgQWer9UJofm-vyGpt_iekLH-ohLrmD0Vnwx3uKla9SGtaQ_V53occk4ZAXIuTRXFFWw'

' Начинаем сбор для игрока: 226: XXWz2wyoT8GTbxNGZ4Bv16D8LLXdagvkxSnOBflUlej2qV787reUFfnrI6B_ru44zGijBwIHthexnA'

' Начинаем сбор для игрока: 227: DvsH3iIc0QjUMoE4LbOf0DgImvJs1MpX0cR6PDxVkqh_R67YGyGcb2wcy-S4Kw-Z1jYZ0HJQbFfr1g'

' Начинаем сбор для игрока: 228: m801SoCWWS_9M3sNsXIzTpEET4MiifwjDdzEpPcnuKDeD7za0G7RQffy2J3TuTVrUVGY7lYgYgy8hA'

' Начинаем сбор для игрока: 229: quMzmIeMHe1F_vG6tw0Az4iOx94NI4X_gEbxSdGEPIUu45Sn0mYcHxRjDNWCVAIIuLLP57MYva1tcQ'

' Начинаем сбор для игрока: 230: OSK0D225_cCo-ft5UE7ca5yOf3Uzsqyb9F9Hxj7nDjC8igA5hIlqyIQlxDqakyx5Ja96pJ4qpV_NLw'

' Начинаем сбор для игрока: 231: fnzkpWWL_vQ3Ha8L2lsHVPTjVtVsiaIMl8_KQgm0cnFB-b0gGgeFJzVAlPC6NUvh9yN52XG4Cy55Dw'

' Начинаем сбор для игрока: 232: x8luYCtu84aVhOQHZJ87XmorIuenLUlusyEqJBrK-PF3Aq4rDrbf4Lyt6khzNj0P0s4BONDdrX3ilw'

' Начинаем сбор для игрока: 233: I6JKllUWZDX4MCYk8jYGNOfcdIGt3gn9BeSXBz89_HDlAaVJHaXpTTL9IqXoyjIYoknrkdLJ67wGew'

' Начинаем сбор для игрока: 234: ytar8eLCBED6jEtiesncmM7Gee4aGRc9MrUlRxzQqMG5e3RBWDMKo0Lhq6VPvG6_yX6LuaJLYI0A9w'

' Начинаем сбор для игрока: 235: 1ZaGt-efB4yVXl8YDw3jyXt9r2zpeIuevJKaxXsw1jfmaTaHQkMBUJs0evolqBflSZPMkrdzXsp_EA'

' Начинаем сбор для игрока: 236: nafiLRGGDdNWsmXh4wSkG74jHX7Y28v3nsr-UZud4CpPNNXdcIX-In_irRbrXsj8J-uJVj1Su32RLQ'

' Начинаем сбор для игрока: 237: AOGRsNuHRhlFhedxQ5lxojxlZEDOzv6v3bpPt80F-6zA6LdMmLyvQmKJ8jXKqKz0b_9e75jjdMUZhA'

' Начинаем сбор для игрока: 238: tZxMSApJWmCCcOt0TGtRs7V3_tjN7lcH2Gt9vub5bGyFvK2aFMKzBnkMlmkfe14-Zob5NwVa-plE_A'

' Начинаем сбор для игрока: 239: vd7O3nHOia-OVND5W8Y8GEu5isFIOBJteHDoiQyv48WciUguGYc2pqECaj1SKgs-Q-SOBdVIY_SOHw'

' Начинаем сбор для игрока: 240: dS7hSZKn9wp6Tf21YjpXZePiCjphJ2-xbJZAXPW_f1yXEd2ORcD0IMDEZsyrWPvJIt81AsknRbFzlQ'

' Начинаем сбор для игрока: 241: ALiXwXF_5zlpyftk5fvKd66RCJXk2gmszlRcgLQdE6uvx7WMF_phI0XcBrGmAl6T0j6llHqw_oZIMA'

' Начинаем сбор для игрока: 242: s7bW_-vbBwIWmzfFozMR63e87uqGaXJ65V1zsS0A6BgbyCa1MlsWBv2RSN5Ogbas8G-DLUIhwIKiRg'

' Начинаем сбор для игрока: 243: Y6d_mey_DHNoBg_qsPS_jqvV1eHFCQta3C84Y-bAM6aCuRjbDPU2P-qqP6pG3I-Qjl8KWo1ddDTMUg'

' Начинаем сбор для игрока: 244: 4oOxU73wfiX6xINEOcjNT2EfsqDc3kFFIApyJyVkWq58rALPbGHCP_tzRvm_A-O6_ipHXJ8hL_vI6Q'

' Начинаем сбор для игрока: 245: tWZGqKZlRN0kuynbbzgI898Zor7lEgnvxQFsHNL2rcRnIb6a_oBRROvDnoo6t55IIXfpKgP4YrPPjw'

' Начинаем сбор для игрока: 246: 96cPlLYR0xSrYpsulBR9cGoCvx2kV3rBaWbVBN3qYsm261ucrzXrmYdKBTlaZ2C3tQlxlFcrswr3GA'

' Начинаем сбор для игрока: 247: 547RYVA_Dwuit9uhVNh0iT6YY_xH8ltwRwRjpUdhzsxysP9IkwIhCDL-bYjiu8C6GhxvrLI2PsZ8Jw'

' Начинаем сбор для игрока: 248: WiGhH3b0aYJThPAzbwSoJMR13I7dwWe_ncSiXF9dVzmwZBk2qkBCUOmsKwiYgthLyesh4-0g2BQQSA'

' Начинаем сбор для игрока: 249: TH0VIEP0Ax8NrdQxnkFY8UjQOqrc1v2p4WmuISp6o_1EtxLznHRAxzqxLRGvbRTBzgpfjarR_NIolg'

' Начинаем сбор для игрока: 250: scyyJ9Dzqd2hAzxCLnN93ipJPdaHZNdIuj5ZZTo6R-2vQwH7ZNaWv-t90r39kJE4IaFkQABfd1XrCg'

' Начинаем сбор для игрока: 251: MUFDAD7W_FA13NdTKMdmymOWCRsgx8hLL0GlIvRCpjDWl9LpS-_9gC8QXfdewHT18O35ML0ltT_I1A'

' Начинаем сбор для игрока: 252: IQPeEiKMWezs85roGyHgbCrA6leJESvbWRQF8pJlB6XDskXajWW9cuAbKA893bYX1sSSVRM37q3UcQ'

' Начинаем сбор для игрока: 253: BOqcSXagaIbKsu8w8yccaIj8KoYx7Dpl3Pgg06eNCoeDeKM31jvU9BDgbY2MC7fUbilfXXoglrQygQ'

' Начинаем сбор для игрока: 254: AZwQz404iTBcr2oJJ132jR4tmJwShFYaJQZAYyZ8bS1mBTPLweUPkvHi_a7r70gYTxr9WQNHHQAlow'

' Начинаем сбор для игрока: 255: A24zbZXSNcLw5azcWyzCqW6NXaFHMo_5d3gK-5T4jWAkaAe_8uEbbw4xNcT8p9PaAPlgpD_R4sXLNw'

' Начинаем сбор для игрока: 256: iRnvp62qo13GGbQUkRko2a5PWaW_r8UFTfalzV9-ZDymWNBxHxtaRLafNFzKDg0c8qyzrk1u9XhkGw'

' Начинаем сбор для игрока: 257: 8Tx22cy4uao7qCIA4Hl7SqdPMR1jq5DagkObBoQNQtJ_2zc9WodlRaSLsDxstpvB56S3Imze8yWpOQ'

' Начинаем сбор для игрока: 258: 4X6zPv-i4y7aYDm2VxxCZ0zfWx2HymKDC74qBW9FT4-886oyCAyOpEppYsrnHZ7KhXrRPe-l9fZCEQ'

' Начинаем сбор для игрока: 259: SqCPv0GvNBJdq5d4vBKfAtquco9tfu0kj-F7M_RdjGn3MZLrDYIvBU7A7U3MlWFNAqKwj5VsZfhW-w'

' Начинаем сбор для игрока: 260: smnKEO1Cctjq1WrIjLAd-zucfa-D4WjhcD2jSiQ4Y0YTaXYG5n8llAMcOhdUL2ge41IAKsjz_HE3Gg'

' Начинаем сбор для игрока: 261: 7LkQUdvXOD_qowHgMO0m6KJuLVKwjhmr4CYkXGca4Ooh6GgPh9e-wR3t0JPiHivbUZaiT6ETXOS1CQ'

' Начинаем сбор для игрока: 262: EJqP5vF9yqJrSyumxxASKPhMv0GRvMuQKtuoCvdX-05T5TPJW1L7BbjsHCNWLeyhSotGmjreJdxB4g'

' Начинаем сбор для игрока: 263: GIOWflDQA-lb5p6zx-kVo9aGZdoXzkK3ZBfz8IngGd5OQVlz2MqPLkP4pbo5Nk1Y_WqcLEtQrfYkaA'

' Начинаем сбор для игрока: 264: rnV9lk_sRHiYAxAEwvoZqAEgk8sEVc41XYu-er1E35uts04gGbWPG2EVhUzY2Gz1gYu-XaD1WzI8qg'

' Начинаем сбор для игрока: 265: bo0-PdVnTG7uiCctVg7Nbo5bO5-nudVhz3lczMP_UKkNe8jVfwFKkCGijv489yeGJ1fPVGyiDzGfPA'

' Начинаем сбор для игрока: 266: hGbwhYKvU0Vuhqaf36KKx4_hr8Rh6oAsUJcXWjMIW2Wt8XnwguuSBobl3G_j72MSEePUmgw3q3rX8Q'

' Начинаем сбор для игрока: 267: g5rLhM94tdYzsBDnbFn7gjTznvKhLtTpQxodPT1n9nkn3t53PLXfs3BNBg2Nodum3yIIgSj1qXU4DQ'

' Начинаем сбор для игрока: 268: AA022ML1zOpjpQm_1DwCgPsvWrd9-PxzXtzmZqSI5wlzqbt20XMLZ6B9EOEZ2XW-26uhRpleE2_HLg'

' Начинаем сбор для игрока: 269: 1oLSZyuZTBK1PhywgfdboWalDkeJjhChbXLFsEfRAw0QmdC5qp977gVJRagVJxSKRUfFX9iAAybzJw'

' Начинаем сбор для игрока: 270: gHJXtiGtsbXw6NKJ5pWIyuKzu1PW_T2uVztdXQnoKfWg5eU1Jzhg-3R4GZOVdoqofVNNuZt2AFSPjA'

' Начинаем сбор для игрока: 271: AbgF2kilBswKZssYm1rrZFRcJCdOqq6grquh4hPa0Ci-piHHb_JuMcoWvdr_K9BjzFVFID_Js-2uMg'

' Начинаем сбор для игрока: 272: zWXBLAdgmGr0Rm4sAMZk3QlzoN_QhzkF2FAEp7iilNzc6kqWyCz68AOqBRSZP6eRgOS9YThayoZpJw'

' Начинаем сбор для игрока: 273: hO7UiSQ61iLZvwGF3kSKDgo78kYJEHJHriCeKhfHly4bnU1SyL4z9Dh1acKDDk_R263_gGjiLquAsA'

' Начинаем сбор для игрока: 274: 3ZALmTCUx5NZ24HwAooZlS8jSOwj35kBM2r9hIZAK0WaWLkJgw-eM1J2Eg505_UOgwfmRxUHclYDvg'

' Начинаем сбор для игрока: 275: fGcxmBV-0GHJFjqbAOZHstSPKTaOOHPN3-VwdMA67GWjWQVpI3dpCsVopaSvyGQaEkRl-uFrm9GplA'

' Начинаем сбор для игрока: 276: iT7XvVpqlqLXz572LcTAKrwpLfE9bN1HoDwgf6HmfDt6Wv_MeYyswFTFt2sC_2SfBa3Xt90otaHUsw'

' Начинаем сбор для игрока: 277: CI6SM4RmHlznPYKBAmWGK0Y7-fCHylsiAF0BQwzVhuOV2zXyEI_wTJqoByZvy3JXlwA3OfRKasBlTw'

' Начинаем сбор для игрока: 278: ARQzyF7qalA5fSqy5s9nGj53X-hmUt9yzz5h9aLOucKMaYqKGfgrs4D8tRE_uIo8vf7DrDzYZcA48w'

' Начинаем сбор для игрока: 279: 1gqChUP3fplkylI_U-6XEM5BCNuc2grFfVtrOAEd2jZqjSo41BYeV9vRHnmdBE-Ig6FKXfKO_2LiwQ'

' Начинаем сбор для игрока: 280: gmvxIUjpzvjeEvNUDfCQV5yNp8dgR5hfTAFQahykYEUZZFfN4WohzIWeFNzdrWibqu6AVGli7Fub2g'

' Начинаем сбор для игрока: 281: Oeh88CRLzbFIX3-EI9KDdb3W-7il-B6Nrdos2XW5RI6NoXAkFkF_Ws9WMljXtFE2wi7vo4TX-VY1Ew'

' Начинаем сбор для игрока: 282: xuhgQz46vS8jv8cDb5-TszW67yjO-p9lFgDfa513GQHNQl3_MHcTWifGUDtPzcHxhLZkUrfy6zLN4A'

' Начинаем сбор для игрока: 283: 0zDp3Mmb_zoUxYb_eftScYKoT2JT3XVxy_C2_VXsbKnpGDOHgTdh-RScVkAeRmxxfUQThUMvvNOhEA'

' Начинаем сбор для игрока: 284: lERVW7B5NTzi72JrDZWw4T1s6DQQ88F8DObxbX6HDLphFUfnZA8YYvtpByTBfPK-cCXnOlI0wrg6yQ'

' Начинаем сбор для игрока: 285: PKkskXxNdzghhqsUe7Z6FYeBMJZoY0RZ2ChgXMW5_g8z_nFKuwtZL-OJEkYLNj4jYeKEw5KzdL70Zg'

' Начинаем сбор для игрока: 286: DfuB7H1jSWaMEq0LbaTg9arkvwmKRVAQD20DztKK5QL3CwKdg9GJRSlqnsJmUVvJH7oWMo3b_dRU8A'

' Начинаем сбор для игрока: 287: Ja9EoxJwwKyhcByYmprpJ3mwEHwl_k3wqKFgS-CDqdw9S_TC6MdDBz2HdNEf0EFZEAosH_B3TWRtUw'

' Начинаем сбор для игрока: 288: kDuK_gDjYEJKKKpnijgYOZcAoEXOkg9Le0UVLE37HtONbxgWJkE7pBL4rfnyWV0LFeI995yTMTBcFg'

' Начинаем сбор для игрока: 289: loSJspjX8gYjopuCY3072u3uejorAy8R_RmyPhs_znPAXIYfb6_KniC9gkgfd97L_aXaCDjPyVNk5g'

' Начинаем сбор для игрока: 290: lsLFz9HJUuOeNRPrx4MIGwoRWSMBc59tDa5G97MozVyrsUMYwsy03npO6ofPturTm4hz5Myyn5Yz7g'

' Начинаем сбор для игрока: 291: jAlXByMosfD_XBVmCby-nYC94LTgjRznb49mCl8npEW_MQBm3m_bg1SpWSyBzxfdurji2c4UK2uynQ'

' Начинаем сбор для игрока: 292: pINSI32I_VqiP2CD0uq929lZpP4iLAaFBCCoSEFWdTxDF6jxP18l5C2Z2Uwmsa76Lw2Gl6pT6DxBJw'

' Начинаем сбор для игрока: 293: BXWbgFcRXBkq1kaTY1PFpqBWXa2KKB_UDEP_TPXSa_B4ViowRc7ZbFEMAnPDQ1wdqi1syaqFcAkBGQ'

' Начинаем сбор для игрока: 294: AOiI0w0CqjK3XoEPkQgAfvX-ePv7b6xr4ineKnQ9BJ3U9mVeqh-i4vMSZss34SPH2qyTTf_uCmQPIw'

' Начинаем сбор для игрока: 295: gwn5_V2kf2nZiG0PzUF2j7XHwbdXh8atNTAbUE_Sq9Y5OPLi4RBuBF4vXIuD0fgD8l4OMTKqIu_yiA'

' Начинаем сбор для игрока: 296: -zB6utkbHZco62V1vFz_TrYuBuSfF4vQOA81oQF-M2aQ53yO6jzuDmBRn6ZTd1NZrLx-fOGdaYpHxg'

' Начинаем сбор для игрока: 297: 7SKhPc0qqviaiR7eGTGyXWFa8FtwTla8vqcEiv9St-tRQrs5VhdToU6DcZcd-gd00SPToZoUuf2Phg'

' Начинаем сбор для игрока: 298: 4c7aS4_QyQTFv9gPe2btNhpkHCA9XQ9vBPCjLWXWDgH04Ts3Q3NVOQ5N7_lEsIaY0Hblg7I9aO4qZw'

' Начинаем сбор для игрока: 299: THUnzdWHKm8Zu_7LjRroBGnCPF6pFJXUPN2E_6QvFwgdNZPk4vlQVmpAEAIvJa3np_BHblo0iBT_Rw'

In [38]:
logger = logging.getLogger()

# Закрываем и удаляем все обработчики, которые держат файлы
for handler in logger.handlers[:]:
    handler.close()  # Закрывает сам файл внутри операционной системы
    logger.removeHandler(handler)  # Отвязывает его от логгера

In [39]:
total_lines = count_lines(output_file_csv)
display(f"Количество собранных матчей по игрокам: {total_lines - 1}")  # Вычитаем строку заголовка

'Количество собранных матчей по игрокам: 28268'

## Выборка уникальных ID матчей за текущий месяц

На предыдущем шаге мы получили файл с ID матчей (до 100 штук) за месяц для самых активных игроков трех лиг и двух регионов. Но, т.к. игроки могут играть друг против друга в одних и тех же матчах и  ID матчей при этом совпадают, оставим только уникальные ID матчей, в дальнейшем получим по ним более детальную информацию.

In [61]:
# файл с матчами
output_file_csv = base_dir /  'players_match_month.csv'

In [62]:
# Открываем файл с ID матчами
try:
    df = pd.read_csv(output_file_csv)
    display(f"Файл загружен успешно: {output_file_csv} | Строк: {df.shape[0]}")
except Exception as e:
    print(f"Ошибка при загрузке: {output_file_csv}: {e}")

'Файл загружен успешно: d:\\datasets\\LOL\\players_match_month.csv | Строк: 28268'

In [63]:
df_unique_matches = pd.DataFrame(df['match_id'].unique(), columns=['match_id'])

In [64]:
print(f'Всего уникальных матчей за месяц у активных игроков: {df_unique_matches.shape[0]}')

Всего уникальных матчей за месяц у активных игроков: 23329


In [65]:
# Сохраним в отдельный файл
output_file_csv = base_dir / "id_unique_matches.csv" 
df_unique_matches.to_csv(output_file_csv, index=False)

## Запрос детальной информации: матчи и игроки

Для этой цели используем эндпоинт **GET /lol/match/v5/matches/{matchId}**, который возвращает информацию о матче по его ID.

**Для матчей** выборочная структура объектов ответа:
- info.gameDuration (long) - длительность матча в секундах
- info.gameCreation (long) - Unix timestamp время, когда игра создана на сервере
- info.gameMode(string) -  режим игры
- info.gameVersion (string) - версия/патч игры

В игре каждый матч — это битва двух команд, где каждый из 10 участников (participants) перед стартом матча выбирает себе одного уникального игрового персонажа, т.н. "чемпиона". 

**Для игроков** выбрираем объекты:
- info.participants.puuid (string) - идентификатор игрока
- info.participants.summonerName (string) - имя - УСТАРЕЛО!!!
- info.participants.kills (int) - убийства
- info.participants.deaths (int) - смерти
- info.participants.assists (int)  - помощь
- info.participants.win (boolean) - 1 - Победа, 0 - Поражение
- info.participants.teamId (int) - 100 (Синяя команда)/200 (Красная команда) 
- info.participants.championId (int)-  уникальный числовой идентификатор используемого чемпиона

Вместо **summonerName** используются **riotIdGameName** (основной игровой ник, виден всем игрокам, может повторяться) и **riotIdTagline** (уникальная метка, позволяет системе отличать игроков с одинаковыми никами). 

Поле **summonerLevel** полностью удалено из истории матчей, уровень постоянно меняется, а матч фиксирует историю. В MATCH-V5 его больше нет.

**Создадим 2 таблицы:**
- Таблица матчей - содержит общую информацию (ID игры, длительность, версия патча).
- Таблица участников (participants): где под одним match_id создается 10 строк (по одной на каждого игрока) с их личной статистикой.

Ниже - скрипт, который берет сохраненные matchId из файла, запрашивает подробные данные по каждому матчу через эндпоинт /lol/match/v5/matches/{matchId} и раскладывает их на три отдельные таблицы: Матчи, Игроки (Участники) и Чемпионы.

Чтобы полностью исключить ошибку Out of Memory, код будет использовать потоковую обработку: парсить по одному матчу за раз, мгновенно записывать новые строки в три CSV-файла в режиме дозаписи (mode='a') и принудительно очищать оперативную память через gc.collect().

In [66]:
input_match_ids_csv = base_dir / "id_unique_matches.csv"   # Файл с ID матчей
output_matches_csv = base_dir / "matches_data.csv"         # Таблица 1: Матчи
output_players_csv = base_dir / "players_data.csv"         # Таблица 2: Участники

In [67]:
# Загрузка исходных ID матчей
if not os.path.exists(input_match_ids_csv):
    display(f"Ошибка! Файл {input_match_ids_csv} не найден.")
    exit()

df_source = pd.read_csv(input_match_ids_csv)
all_match_ids =  set(df_source['match_id'].dropna())
display(f"Загружено уникальных матчей для обработки: {len(all_match_ids)}")

'Загружено уникальных матчей для обработки: 23329'

In [68]:
# 1. Загрузка исходных ID
if not os.path.isfile(input_match_ids_csv):
    raise FileNotFoundError(f"Исходный файл не найден: {input_match_ids_csv}")

try:
    df_source = pd.read_csv(input_match_ids_csv, usecols=['match_id'])
    unique_match_ids = set(df_source['match_id'].dropna().unique())
    print(f"Всего исходных матчей: {len(unique_match_ids)}")
except KeyError:
    raise KeyError("В исходном файле отсутствует колонка 'match_id'")

# 2. Фильтрация уже обработанных ID
processed_ids = set()
if os.path.isfile(output_matches_csv) and os.path.getsize(output_matches_csv) > 0:
    try:
        df_processed = pd.read_csv(output_matches_csv, usecols=['match_id'])
        processed_ids = set(df_processed['match_id'].dropna().unique())
        print(f"Уже обработано ранее: {len(processed_ids)}")
    except KeyError:
        print("В файле результатов нет 'match_id'. Обработка начнется с нуля.")

# 3. Вычисление остатка
unique_match_ids = list(all_match_ids - processed_ids)
print(f"Осталось обработать пропущенных матчей: {len(unique_match_ids)}")

if not unique_match_ids:
    print("Все матчи уже успешно обработаны!")

Всего исходных матчей: 23329
Уже обработано ранее: 19585
Осталось обработать пропущенных матчей: 12386


In [69]:
def save_batch_to_csv(batch_matches, batch_players):
    """
    Принимает списки словарей для матчей и игроков, 
    переводит в DataFrame и дописывает в CSV.
    """
    if not batch_matches or not batch_players:
        return

    # 1. Сохраняем батч матчей
    df_matches = pd.DataFrame(batch_matches)
    matches_exist = os.path.isfile(output_matches_csv)
    df_matches.to_csv(output_matches_csv, mode='a', index=False, header=not matches_exist, encoding='utf-8')

    # 2. Сохраняем батч игроков (здесь будет ~1000 строк)
    df_players = pd.DataFrame(batch_players)
    players_exist = os.path.isfile(output_players_csv)
    df_players.to_csv(output_players_csv, mode='a', index=False, header=not players_exist, encoding='utf-8')

    print(f"💾 Батч успешно записан на диск ({len(batch_matches)} матчей, {len(batch_players)} игроков).")


In [70]:
def save_match_batch_to_csv(batch_list, file_path="matches_data.csv"):
    """
    Принимает список из 100 JSON-объектов матчей, 
    выпрямляет их и дописывает в CSV-файл.
    """
    if not batch_list:
        return
        
    all_participants = []
    
    # 1. Разворачиваем вложенный JSON матчей для CSV
    for match in batch_list:
        match_id = match['metadata']['matchId']
        game_version = match['info']['gameVersion']
        
        # Вытаскиваем данные всех 10 игроков из этого матча
        for participant in match['info']['participants']:
            # Создаем плоский словарь для строки CSV
            player_row = {
                'match_id': match_id,
                'game_version': game_version,
                'puuid': participant['puuid'],
                'champion_name': participant['championName'],
                'win': participant['win'],  # Наше булево поле!
                'kills': participant['kills'],
                'deaths': participant['deaths'],
                'assists': participant['assists']
                # Добавьте сюда любые другие нужные вам поля
            }
            all_participants.append(player_row)
            
    # 2. Переводим накопленный батч в DataFrame
    df_batch = pd.DataFrame(all_participants)
    
    # 3. Записываем в файл в режиме добавления (mode='a')
    # Если файл еще не создан, пишем его с заголовками (header=True)
    file_exists = os.path.isfile(file_path)
    
    df_batch.to_csv(file_path, mode='a', index=False, header=not file_exists, encoding='utf-8')
    print(f"💾 Батч успешно дозаписан в CSV. Файл: {file_path}")


In [71]:
def match_data_collection(match_id, match_index):
    """
    Запрашивает один матч, парсит его на 2 структуры и возвращает их без записи на диск.
    """
    region_prefix = match_id.split('_')[0].lower()
    route = get_route(region_prefix) 
    url = f"https://{route}.api.riotgames.com/lol/match/v5/matches/{match_id}"
    
    try:
        response = requests.get(url, headers=HEADERS, timeout=5)
        
        if response.status_code == 429:
            print("⚠️ Лимит запросов! Засыпаем на 12 секунд...")
            time.sleep(12)
            return None
            
        if not response.ok:
            print(f"❌ Ошибка API для матча {match_id}: Код {response.status_code}")
            return None

        match_data = response.json()
        info = match_data.get("info", {})
        
        if not info:
            return None

        # --- 1. ПАРСИНГ ДАННЫХ МАТЧА ---
        match_row = {
            "match_id": match_id,
            "game_name":  info.get("gameName"), 
            "game_creation": info.get("gameCreation"),
            "game_duration": info.get("gameDuration"),
            "game_version": info.get("gameVersion"),
            "game_mode": info.get("gameMode"),
            "endOfGameResult": info.get("endOfGameResult"),
            "gameStartTimestamp": info.get("gameStartTimestamp"),
            "gameEndTimestamp": info.get("gameEndTimestamp"),
            "gameType": info.get("gameType"),
            "platform_id" : info.get("platformId"),
            "queue_id": info.get("queueId")
        }

        # --- 2. ПАРСИНГ ИГРОКОВ (10 участников на матч) ---
        match_players = []
        participants = info.get("participants", [])
        for p in participants:
            player_row = {
                "match_id": match_id,
                "puuid": p.get("puuid"),
                #"summoner_name": p.get("summonerName"),   # Устарело
                #"summoner_level": p.get("summonerLevel"), # Устарело
                "riotIdGameName" : p.get("riotIdGameName"),
                "riotIdTagline" :  p.get("riotIdTagline"),
                "kills": p.get("kills"),
                "deaths": p.get("deaths"),
                "assists": p.get("assists"),
                "win": p.get("win"),
                "teamPosition": p.get("teamPosition"),
                "goldEarned": p.get("goldEarned"),
                "goldSpent": p.get("goldSpent"),
                "team_id": p.get("teamId"),
                "champion_id": p.get("championId"),
                "championName":  p.get("championName"),
                "timePlayed": p.get("timePlayed")
            }
            match_players.append(player_row)

        print(f"👁️ Матч {match_index} - {match_id} обработан локально.")
        
        # Возвращаем кортеж: строка матча и список строк его игроков
        return match_row, match_players

    except Exception as e:
        print(f"❌ Критическая ошибка при обработке матча {match_id}: {e}")
        return None


In [72]:
# --- ГЛАВНЫЙ ПОТОКОВЫЙ ИТЕРАТОР (батчи по 100 штук)
def main_pipeline():
    print("🚀 Запуск потокового парсинга матчей пачками по 100 штук...")
    
    batch_size = 100
    accumulated_matches = []
    accumulated_players = []
    
    for index, match_id in enumerate(unique_match_ids, start=1):
        match_id = str(match_id).strip()
        if not match_id:
            continue
               
        # Вызываем парсер
        result = match_data_collection(match_id, index)
            
        if result:
            match_row, match_players = result
            accumulated_matches.append(match_row)
            accumulated_players.extend(match_players) # Разворачиваем список из 10 игроков
            
        # Если набрали 100 матчей ИЛИ это конец списка уникальных ID
        if len(accumulated_matches) == batch_size or index == len(unique_match_ids):
            print(f"📦 Накоплен батч! Запись {len(accumulated_matches)} матчей...")
            
            # Записываем пачку на диск за один раз
            save_batch_to_csv(accumulated_matches, accumulated_players)
            
            # Очищаем списки и принудительно освобождаем RAM
            accumulated_matches.clear()
            accumulated_players.clear()
            gc.collect() 
            
        # Rate Limit Riot API (1.2 сек — надежный тайминг для 1 запроса)
        time.sleep(1.2)
        
    print("🏁 Весь процесс парсинга успешно завершен!")


In [73]:
main_pipeline()
display("Сбор данных закончен")

🚀 Запуск потокового парсинга матчей пачками по 100 штук...
👁️ Матч 1 - NA1_5562266569 обработан локально.
👁️ Матч 2 - NA1_5568787128 обработан локально.
👁️ Матч 3 - NA1_5570800772 обработан локально.
👁️ Матч 4 - NA1_5568372832 обработан локально.
👁️ Матч 5 - EUW1_7862698204 обработан локально.
👁️ Матч 6 - NA1_5567754321 обработан локально.
👁️ Матч 7 - NA1_5570698001 обработан локально.
👁️ Матч 8 - EUW1_7871699959 обработан локально.
👁️ Матч 9 - NA1_5570886685 обработан локально.
👁️ Матч 10 - EUW1_7869785584 обработан локально.
👁️ Матч 11 - NA1_5565685202 обработан локально.
👁️ Матч 12 - EUW1_7867066341 обработан локально.
👁️ Матч 13 - NA1_5563167505 обработан локально.
👁️ Матч 14 - NA1_5568647804 обработан локально.
👁️ Матч 15 - NA1_5572341321 обработан локально.
👁️ Матч 16 - NA1_5564207578 обработан локально.
👁️ Матч 17 - NA1_5570654226 обработан локально.
👁️ Матч 18 - EUW1_7861001289 обработан локально.
👁️ Матч 19 - EUW1_7859055947 обработан локально.
👁️ Матч 20 - EUW1_7852160968 обр

'Сбор данных закончен'

## Вывод

В период 14.06.2026-16.06.2026 с помощью официального API Riot Games были собраны следующие данные за май 2026:

- **all_players_data.csv** - 21631 записей об игроках в лигах Challenger, Grandmaster, Master для регионов euw1 и na1 с актуальными данными по игроку на 15.06.2026
  
- из all_players_data.csv выбрано по 50 самых активных игроков по каждой лиге и региону - получен файл **top_players.csv** (300 записей)

- для каждого из 300 игроков из top_players.csv получен список до 100 матчей, в которых игрок участвовал в мае 2026 года - **players_match_month.csv** (28269 записей)

- для каждого матча из players_match_month.csv с помощью API получена детальная информация о матчах - **matches_data.csv** (31965 записей)

- для каждого матча из matches_data.csv получены данные по игрокам, участвующих в матчах -  **players_data.csv** (343524 записи)